
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Comparing proportions &#8212; The Python Companion of Intuitive Biostatistics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=G-QQY9FLLPJ8"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'G-QQY9FLLPJ8');
            </script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'G-QQY9FLLPJ8');
            </script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '27 - Comparing proportions';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Comparing survival curves" href="29%20-%20Comparing%20survival%20curves.html" />
    <link rel="prev" title="Normality tests and outliers" href="24%20-%20Normality%20tests%20and%20outliers.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.jpg" class="logo__image only-light" alt="The Python Companion of Intuitive Biostatistics - Home"/>
    <script>document.write(`<img src="_static/logo.jpg" class="logo__image only-dark" alt="The Python Companion of Intuitive Biostatistics - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="README.html">
                    Python Companion Guide to “Intuitive Biostatistics”
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Continuous variables</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="09%20-%20Quantifying%20scatter%20of%20continuous%20data.html">Quantifying scatter of continuous data</a></li>
<li class="toctree-l1"><a class="reference internal" href="10%20-%20Gaussian%20distribution.html">The Gaussian distribution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Confidence intervals</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="12%20-%20Confidence%20interval%20of%20a%20mean.html">Confidence interval of a mean</a></li>
<li class="toctree-l1"><a class="reference internal" href="04%20-%20Confidence%20interval%20of%20a%20proportion.html">Confidence interval of a proportion</a></li>
<li class="toctree-l1"><a class="reference internal" href="05%20-%20Confidence%20interval%20of%20survival%20data.html">Confidence interval of survival data</a></li>
<li class="toctree-l1"><a class="reference internal" href="06%20-%20Confidence%20interval%20of%20counted%20data.html">Confidence interval of counted data</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical significance and data assumptions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="15%20-%20Statistical%20significance.html">Statistical significance</a></li>
<li class="toctree-l1"><a class="reference internal" href="20%20-%20Statistical%20power%20and%20sample%20size.html">Statistical power and sample size</a></li>
<li class="toctree-l1"><a class="reference internal" href="24%20-%20Normality%20tests%20and%20outliers.html">Normality tests and outliers</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical tests</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Comparing proportions</a></li>
<li class="toctree-l1"><a class="reference internal" href="29%20-%20Comparing%20survival%20curves.html">Comparing survival curves</a></li>
<li class="toctree-l1"><a class="reference internal" href="30%20-%20Comparing%20two%20unpaired%20means.html">Comparing two unpaired means</a></li>
<li class="toctree-l1"><a class="reference internal" href="31%20-%20Comparing%20paired%20data.html">Comparing paired data</a></li>
<li class="toctree-l1"><a class="reference internal" href="32%20-%20Correlation.html">Correlation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Fitting models to data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="33%20-%20Simple%20linear%20regression.html">Simple linear regression</a></li>
<li class="toctree-l1"><a class="reference internal" href="35%20-%20Comparing%20models.html">Comparing models</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/sbwiecko/intuitive_biostatistics" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sbwiecko/intuitive_biostatistics/edit/master/27 - Comparing proportions.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/27 - Comparing proportions.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Comparing proportions</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#fisher-s-exact-test">Fisher’s exact test</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#when-to-use-fisher-s-test">When to use Fisher’s test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#contingency-table">Contingency table</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#p-value">P value</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#hypergeometric-distribution">Hypergeometric distribution</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#relationship-between-hypergeometric-distribution-and-p-value">Relationship between hypergeometric distribution and P value</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-p-value">Interpreting the P value</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-and-interpreting-risk-metrics">Calculating and interpreting risk metrics</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#absolute-risk-reduction-arr-and-attributable-risk-ar">Absolute risk reduction (ARR) and attributable risk (AR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-attributable-risk-par">Population attributable risk (PAR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#number-needed-to-treat-ntt">Number needed to treat (NTT)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#relative-risk-rr">Relative risk (RR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#logrr">LogRR</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#odds-ratio-or">Odds ratio (OR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#logor">LogOR</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#confidence-intervals">Confidence intervals</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-proportions">CI for proportions</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-arr">CI for ARR</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#standard-approximations">Standard approximations</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#z-test-for-comparing-proportions">Z-test for comparing proportions</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#exact-methods">Exact methods</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-nnt">CI for NNT</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-rr">CI for RR</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-or">CI for OR</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-table">Summary table</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-measures-of-association-and-effect-with-custom-functions">Calculating measures of association and effect with custom functions</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#table2x2-for-risk-and-odds-ratio-analysis">Table2x2 for risk and odds ratio analysis</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#one-sided-test-application-to-toxicology-count-data">One-sided test - Application to toxicology count data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-test">Chi-squared test</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-it-works">How it works</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-of-the-mendel-s-peas">Example of the Mendel’s peas</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-goodness-of-fit-test">Chi-squared goodness-of-fit test</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-chi-squared-distribution">The chi-squared distribution</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#degrees-of-freedom-in-goodness-of-fit-tests">Degrees of freedom in goodness-of-fit tests</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-test-of-independence">Chi-squared test of independence</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mosaic-plot">Mosaic plot</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-results-of-the-chi-squared-test-of-independence">Interpreting the results of the chi-squared test of independence</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#degrees-of-freedom-in-independence-tests">Degrees of freedom in independence tests</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#yates-correction">Yates’ correction</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#directional-hypothesis">Directional hypothesis</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-independence-tests-using-pingouin">Chi-squared independence tests using Pingouin</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cochran-armitage-test-for-trend">Cochran-Armitage test for trend</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-the-trend-test-works">How the trend test works</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-of-varying-the-dose-of-a-drug">Example of varying the dose of a drug</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#table-vs-table2x2"><code class="docutils literal notranslate"><span class="pre">Table</span></code> vs. <code class="docutils literal notranslate"><span class="pre">Table2x2</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-results-of-the-cochran-armitage-test">Interpreting the results of the Cochran-Armitage test</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#case-control-studies">Case-control studies</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-analysis-in-case-control-studies">Statistical analysis in case-control studies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cholera-vaccination-example">Cholera vaccination example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#isotretinoin-and-bowel-disease">Isotretinoin and bowel disease</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">Conclusion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cheat-sheet">Cheat sheet</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Contingency table</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Fisher’s exact test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#risk-and-odds-ratio-analysis">Risk and odds ratio analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#z-test">Z-test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Chi-squared goodness-of-fit test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-distribution">Chi-squared distribution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Chi-squared test of independence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Cochran-Armitage test for trend</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Case-control studies</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-information">Session information</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="comparing-proportions">
<h1>Comparing proportions<a class="headerlink" href="#comparing-proportions" title="Link to this heading">#</a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>This chapter embarks on an exploration of the methodologies and Python tools that empower us to analyze and interpret scenarios where outcomes are classified into <strong>categories</strong>. Whether we’re investigating the prevalence of a disease, the effectiveness of a treatment, or the association between risk factors and outcomes, the comparison of proportions provides a statistical lens through which we can discern meaningful relationships.</p>
<p>Proportions, representing the <em>fraction of individuals or events falling into specific categories</em>, offer a concise and intuitive way to summarize categorical data. The comparison of proportions extends this simplicity to uncover potential disparities or associations. By examining the <strong>differences or ratios</strong> between proportions derived from distinct groups or conditions, we gain insights into:</p>
<ul class="simple">
<li><p>Disease prevalence: how does the occurrence of a condition vary across populations or over time?</p></li>
<li><p>Treatment efficacy: does a new intervention lead to a higher proportion of successful outcomes compared to a standard treatment or a control group?</p></li>
<li><p>Risk factor analysis: is there a correlation between exposure to certain factors and the likelihood of developing a specific health outcome?</p></li>
<li><p>Trend assessment: does the proportion of individuals exhibiting a particular trait change systematically across ordered categories (e.g., dosage levels in toxicology)?</p></li>
</ul>
<p>We will go through a versatile toolkit for comparing proportions, leveraging the capabilities of Python to perform robust analyses:</p>
<ul class="simple">
<li><p>Contingency tables: the foundation for <strong>organizing and visualizing</strong> categorical relationships.</p></li>
<li><p>Fisher’s Exact test: the go-to for <em>small sample sizes</em> when comparing proportions.</p></li>
<li><p>Chi-Squared test: a powerful workhorse for <em>larger samples</em>, assessing the significance of observed differences.</p></li>
<li><p>Cochran-Armitage trend test: a specialized tool for detecting trends in proportions across <strong>ordered categories</strong> - particularly relevant in toxicology studies.</p></li>
<li><p>Case-control studies: a retrospective approach that investigates if past exposures are more common among individuals with a disease compared to those without.</p></li>
</ul>
</section>
<section id="fisher-s-exact-test">
<h2>Fisher’s exact test<a class="headerlink" href="#fisher-s-exact-test" title="Link to this heading">#</a></h2>
<p>The core idea of Fisher’s test is to calculate the probability of obtaining the exact distribution of the data (or one more extreme) if there were truly no association between the variables (the null hypothesis H0). It does this by considering all possible ways the data could be arranged while keeping the row and column totals fixed. This allows for a precise P value calculation, even when other tests might falter.</p>
<section id="when-to-use-fisher-s-test">
<h3>When to use Fisher’s test<a class="headerlink" href="#when-to-use-fisher-s-test" title="Link to this heading">#</a></h3>
<p>When dealing with categorical data in biostatistics, we often find ourselves in situations where sample sizes are limited. This is where Fisher’s exact test shines. Unlike other tests that rely on approximations (which may not be accurate with small samples), Fisher’s test provides an exact calculation of the probability of observing the data (or something more extreme) under the null hypothesis of no association.</p>
<p>Fisher’s test is particularly well-suited for:</p>
<ul class="simple">
<li><p>Small sample sizes: if the contingency table has cells with expected counts less than 5, Fisher’s test is often the preferred choice.</p></li>
<li><p>Rare events: when we’re dealing with outcomes that occur infrequently, Fisher’s test is more reliable than tests that rely on large-sample assumptions.</p></li>
<li><p>2x2 tables: while Fisher’s test can be extended to larger tables, it’s most commonly used for situations with two categorical variables, each having two levels (e.g., treatment vs. control, success vs. failure).</p></li>
</ul>
</section>
<section id="contingency-table">
<h3>Contingency table<a class="headerlink" href="#contingency-table" title="Link to this heading">#</a></h3>
<p>A contingency table is a structured way to organize the results of a study where both the predictor (e.g., treatment or exposure) and the outcome are categorical. Rows typically correspond to the different levels of the predictor, while columns represent the different possible outcomes. The values within the table show the <strong>count</strong> of subjects falling into each combination of predictor and outcome. We can use <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/stats.contingency.html"><code class="docutils literal notranslate"><span class="pre">scipy.stats.contingency</span></code> functions</a> or <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html"><code class="docutils literal notranslate"><span class="pre">pandas.crosstab</span></code></a> to create and analyze contingency tables.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">contingency</span>

<span class="c1"># Example data</span>
<span class="n">treatment</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span> <span class="c1"># 1 = Drug A, 2 = Drug B</span>
<span class="n">outcome</span>   <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span> <span class="c1"># 0 = No response, 1 = Response</span>


<span class="c1"># Create contingency table with scipy.stats.contingency</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">contingency</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">treatment</span><span class="p">,</span> <span class="n">outcome</span><span class="p">)</span> <span class="c1"># treatment will be the rows, outcome the columns</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;With `scipy.stats.contingency`&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;------------------------------&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The object containing all atributes is:</span><span class="se">\n</span><span class="si">{</span><span class="n">res</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The `treatment` (rows) elements are: </span><span class="si">{</span><span class="n">res</span><span class="o">.</span><span class="n">elements</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The `outcome` (columns) elements are: </span><span class="si">{</span><span class="n">res</span><span class="o">.</span><span class="n">elements</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The actual contingency table is:</span><span class="se">\n</span><span class="si">{</span><span class="n">res</span><span class="o">.</span><span class="n">count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># If the data is already in a Pandas DataFrame we can use pandas.crosstab</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Assuming the Dataframe has columns named &quot;treatment&quot; and &quot;outcome&quot;</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;treatment&quot;</span><span class="p">:</span> <span class="n">treatment</span><span class="p">,</span> <span class="s2">&quot;outcome&quot;</span><span class="p">:</span> <span class="n">outcome</span><span class="p">})</span>


<span class="c1"># Creating contingency table using pd.crosstab</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;treatment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">(),</span>  <span class="c1"># Remove NaN values if expected</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;outcome&quot;</span><span class="p">],</span>
    <span class="n">margins</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>     <span class="c1"># Include row and column totals</span>
    <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span>      <span class="c1"># Keep rows/columns with zero counts</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">With `pandas.crosstab`&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;----------------------&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>With `scipy.stats.contingency`
------------------------------
The object containing all atributes is:
CrosstabResult(elements=(array([1, 2]), array([0, 1])), count=array([[2, 2],
       [1, 2]]))
The `treatment` (rows) elements are: [1 2]
The `outcome` (columns) elements are: [0 1]
The actual contingency table is:
[[2 2]
 [1 2]]

With `pandas.crosstab`
----------------------
outcome    0  1  All
treatment           
1          2  2    4
2          1  2    3
All        3  4    7
</pre></div>
</div>
</div>
</div>
<p>Let’s prepare the data derived from the study from <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/23216615/">Agnelli and colleagues</a>, and where each value is an actual number of patients, in a contingency table.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Treatment</p></th>
<th class="head"><p>Recurrent VTE</p></th>
<th class="head"><p>No recurrence</p></th>
<th class="head"><p>Total</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Placebo</p></td>
<td><p>73</p></td>
<td><p>756</p></td>
<td><p>829</p></td>
</tr>
<tr class="row-odd"><td><p>Apixaban, 2.5 mg</p></td>
<td><p>14</p></td>
<td><p>826</p></td>
<td><p>840</p></td>
</tr>
<tr class="row-even"><td><p>Total</p></td>
<td><p>87</p></td>
<td><p>1582</p></td>
<td><p>1669</p></td>
</tr>
</tbody>
</table>
</div>
<p>The goal of this analysis is to draw conclusions about the <em>general population</em> of patients who have experienced a thromboembolism. This is achieved by calculating confidence intervals (CIs) and a P value. The P value helps us answer a critical question: if the treatment truly doesn’t affect the risk of recurrence (our null hypothesis, H0), what’s the probability of observing incidence rates as different (or even more different) than what we see in our study, simply due to random chance?</p>
<p>The most appropriate statistical test for this situation is Fisher’s exact test. However, Fisher’s test can become computationally challenging with large sample sizes. In those cases, the chi-square test (as discussed later) serves as a practical alternative, yielding nearly identical P values. The Fisher’s exact test is performed with <a class="reference external" href="https://docs.scipy.org/doc/scipy-1.13.1/reference/generated/scipy.stats.fisher_exact.html"><code class="docutils literal notranslate"><span class="pre">scipy.stats.fisher_exact</span></code></a> on a 2x2 contingency table.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Contingency table with rows = alternative treatments, cols = alternative outcomes</span>
<span class="c1"># table = np.array([[73, 756], [14, 826]])</span>

<span class="n">table_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>   <span class="c1"># pre-aggregated data</span>
        <span class="s1">&#39;recurrence&#39;</span><span class="p">:</span>    <span class="p">[</span><span class="mi">73</span><span class="p">,</span> <span class="mi">14</span><span class="p">],</span>
        <span class="s1">&#39;no_recurrence&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">756</span><span class="p">,</span> <span class="mi">826</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;placebo&#39;</span><span class="p">,</span> <span class="s1">&#39;apixaban&#39;</span><span class="p">],</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">table_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>          recurrence  no_recurrence
placebo           73            756
apixaban          14            826
</pre></div>
</div>
</div>
</div>
</section>
<section id="p-value">
<h3>P value<a class="headerlink" href="#p-value" title="Link to this heading">#</a></h3>
<section id="hypergeometric-distribution">
<h4>Hypergeometric distribution<a class="headerlink" href="#hypergeometric-distribution" title="Link to this heading">#</a></h4>
<p>Fisher’s exact test relies on the <a class="reference external" href="https://en.wikipedia.org/wiki/Hypergeometric_distribution">hypergeometric distribution</a> to compute the exact probability of observing a given contingency table (or one more extreme) under the null hypothesis of no association. Briefly, the hypergeometric distribution describes the probability of obtaining a specific number of successes in a sample drawn without replacement from a finite population containing two types of objects (e.g., cases and controls, exposed and unexposed).</p>
<p>Assume the null hypothesis of independence is true, and constrain the marginal counts to be as observed. What is the chance of getting this exact table and what is the chance of getting a table at least as extreme?</p>
<p>Imagine an urn with n white balls and <em>M - n</em> black balls. Draw k balls <em>without replacement</em>, and let k be the number of white balls in the sample:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head text-center"><p>white</p></th>
<th class="head text-center"><p>black</p></th>
<th class="head text-center"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sampled</p></td>
<td class="text-center"><p>k</p></td>
<td class="text-center"><p>N - k</p></td>
<td class="text-center"><p>N</p></td>
</tr>
<tr class="row-odd"><td><p>not sampled</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>M - N</p></td>
</tr>
<tr class="row-even"><td><p></p></td>
<td class="text-center"><p>n</p></td>
<td class="text-center"><p>M - n</p></td>
<td class="text-center"><p>M</p></td>
</tr>
</tbody>
</table>
</div>
<p>k follows a hypergeometric distribution <span class="math notranslate nohighlight">\(K \sim \text{Hypergeometric}(M, n, N)\)</span>, with</p>
<div class="math notranslate nohighlight">
\[
P(K=k) = \frac{\binom{n}{k} \binom{M - n}{N - k}}{\binom{M}{N}}
\]</div>
<p>For example, we can use the probability mass function (PMF) from <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.hypergeom.html"><code class="docutils literal notranslate"><span class="pre">scipy.stats.hypergeom</span></code></a> to solve the following configuration:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head text-center"><p>0</p></th>
<th class="head text-center"><p>1</p></th>
<th class="head text-center"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sampled</p></td>
<td class="text-center"><p>18</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>20</p></td>
</tr>
<tr class="row-odd"><td><p>not sampled</p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p></p></td>
<td class="text-center"><p>20</p></td>
</tr>
<tr class="row-even"><td><p></p></td>
<td class="text-center"><p>29</p></td>
<td class="text-center"><p>11</p></td>
<td class="text-center"><p>40</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">hypergeom</span>

<span class="n">k</span><span class="p">,</span> <span class="n">M</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">N</span> <span class="o">=</span> <span class="mi">18</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">29</span><span class="p">,</span> <span class="mi">20</span>
<span class="n">pmf_hypergeom</span> <span class="o">=</span> <span class="n">hypergeom</span><span class="o">.</span><span class="n">pmf</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">M</span><span class="o">=</span><span class="n">M</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;P(K=18) = </span><span class="si">{</span><span class="n">pmf_hypergeom</span><span class="si">:</span><span class="s2">.5</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>P(K=18) = 0.013804
</pre></div>
</div>
</div>
</div>
</section>
<section id="relationship-between-hypergeometric-distribution-and-p-value">
<h4>Relationship between hypergeometric distribution and P value<a class="headerlink" href="#relationship-between-hypergeometric-distribution-and-p-value" title="Link to this heading">#</a></h4>
<p>In the context of Fisher’s test, the “successes” are the individuals with the outcome of interest (e.g., cases in a case-control study), and the “population” is the total number of individuals in the study. The row and column totals in the contingency table define the number of successes and the sample sizes in each group. The probability of observing a particular contingency table with values a, b, c, and d is given by the hypergeometric probability mass function:</p>
<div class="math notranslate nohighlight">
\[P(a) = \frac{\binom{a+b}{a} \binom{c+d}{c}}{\binom{a+b+c+d}{a+c}}\]</div>
<p>The P value for Fisher’s exact test of independence can be obtained using either the cumulative distribution function (CDF) or the survival function (SF) of the hypergeometric distribution, <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html">as explained in greater details elsewhere</a>, depending on the directionality of the test:</p>
<ul class="simple">
<li><p>One-tailed test (right tail): when the alternative hypothesis suggests a higher number of successes than expected, the P value is calculated using the survival function (SF), which gives the probability of observing a value greater than or equal to the observed value: <span class="math notranslate nohighlight">\(P = P(K \ge k) = \text{SF}(k - 1)\)</span>, where <span class="math notranslate nohighlight">\(K\)</span> is the random variable representing the number of successes, and <span class="math notranslate nohighlight">\(k\)</span> is the observed number of successes.</p></li>
<li><p>One-tailed test (left tail): when the alternative hypothesis suggests a lower number of successes than expected, the P value is calculated using the cumulative distribution function (CDF), which gives the probability of observing a value less than or equal to the observed value: <span class="math notranslate nohighlight">\(P = P(K \le k) = \text{CDF}(k)\)</span></p></li>
<li><p>Two-tailed test: in a two-tailed test, we typically consider deviations in both directions (<em>more extreme associations</em> in either direction) from the observed table. Therefore, to obtain the two-sided P value, we often double the P value obtained from the one-sided calculation (using either the CDF or SF, depending on which tail is more extreme).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">fisher_exact</span><span class="p">,</span> <span class="n">hypergeom</span>

<span class="c1"># Contingency table </span>
<span class="n">table</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">18</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">11</span><span class="p">,</span> <span class="mi">9</span><span class="p">]]</span>

<span class="c1"># Perform Fisher&#39;s exact test</span>
<span class="n">odds_ratio</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;two-sided&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fisher&#39;s exact test (two-Sided); P value: </span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Parameters for the hypergeometric distribution (from contingency table)</span>
<span class="n">M</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">table</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">table</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">n</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">table</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">N</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># 1. Calculate one-sided P value (more extreme in one direction)</span>
<span class="n">k</span><span class="o">=</span><span class="mi">18</span>  <span class="c1"># number of successes</span>
<span class="n">p_value_one_sided</span> <span class="o">=</span> <span class="n">hypergeom</span><span class="o">.</span><span class="n">sf</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="n">k</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">M</span><span class="o">=</span><span class="n">M</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;One-sided P value (&gt;= </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> &#39;successes&#39;): </span><span class="si">{</span><span class="n">p_value_one_sided</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># 2. Calculate two-sided P value (more extreme in either direction)</span>
<span class="n">p_value_two_sided</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">p_value_one_sided</span>  <span class="c1"># Assuming symmetry</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Two-sided P value: </span><span class="si">{</span><span class="n">p_value_two_sided</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># 3. Probabilities of individual tables</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Probabilities of individual tables (more extreme than observed):&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">18</span><span class="p">,</span> <span class="mi">19</span><span class="p">,</span> <span class="mi">20</span><span class="p">]:</span>  <span class="c1"># Consider tables with more extreme &#39;successes&#39;</span>
    <span class="n">prob</span> <span class="o">=</span> <span class="n">hypergeom</span><span class="o">.</span><span class="n">pmf</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">M</span><span class="o">=</span><span class="n">M</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  k = </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="si">{</span><span class="n">prob</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fisher&#39;s exact test (two-Sided); P value: 0.0310
One-sided P value (&gt;= 18 &#39;successes&#39;): 0.0155
Two-sided P value: 0.0310
Probabilities of individual tables (more extreme than observed):
  k = 9:	0.0001
  k = 10:	0.0016
  k = 11:	0.0138
  k = 18:	0.0138
  k = 19:	0.0016
  k = 20:	0.0001
</pre></div>
</div>
</div>
</div>
</section>
<section id="interpreting-the-p-value">
<h4>Interpreting the P value<a class="headerlink" href="#interpreting-the-p-value" title="Link to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assuming &#39;table&#39; is our 2x2 contingency table (NumPy array or list of lists)</span>
<span class="c1"># For pre-aggregated data we can do:</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span> <span class="c1"># converts DataFrame to NumPy array</span>
<span class="n">odds_ratio</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;P value in the apixaban study: </span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s2">.3e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>P value in the apixaban study: 1.334e-11
</pre></div>
</div>
</div>
</div>
<p>A very low P value typically leads to the rejection of the null hypothesis. It’s unlikely that the observed results (the distribution of data in the contingency table) occurred purely by chance if there were truly no association between the variables. Said differently, the low P value indicates that there’s a statistically significant <strong>association</strong> between the two categorical variables.</p>
</section>
</section>
<section id="calculating-and-interpreting-risk-metrics">
<h3>Calculating and interpreting risk metrics<a class="headerlink" href="#calculating-and-interpreting-risk-metrics" title="Link to this heading">#</a></h3>
<section id="absolute-risk-reduction-arr-and-attributable-risk-ar">
<h4>Absolute risk reduction (ARR) and attributable risk (AR)<a class="headerlink" href="#absolute-risk-reduction-arr-and-attributable-risk-ar" title="Link to this heading">#</a></h4>
<p>In the context of evaluating treatments, the <strong>absolute risk reduction (ARR)</strong>, also known as <strong>attributable risk (AR)</strong> or risk difference, is a key measure that quantifies the absolute difference in the risk of an outcome between the treatment group and the control (or placebo) group. It provides a direct estimate of the reduction in risk that can be attributed to the treatment. The foundation of ARR is the concept of <strong>risk</strong>, which is simply the probability of an event (e.g., disease, side effect, recovery) occurring within a specific time period. Risk is often expressed as a percentage or a decimal:</p>
<div class="math notranslate nohighlight">
\[\text{Risk} = p = \frac{\text{Number of events in the group}}{\text{Total number of individuals in the group}}\]</div>
<p>While ARR and AR are mathematically equivalent, their usage and interpretation differ slightly:</p>
<ul class="simple">
<li><p>ARR: typically used when the <em>exposure or intervention is beneficial</em>, emphasizing the positive impact of reducing the risk of an undesirable outcome. The control or placebo group will be in the first row of the contingency table.</p></li>
<li><p>AR : more commonly used when the <em>exposure is harmful</em>, highlighting the excess risk associated with that exposure. The exposed group will be in the first row of the contingency table, like in the example of the oral contraceptive study discussed later.</p></li>
</ul>
<p>We can calculate ARR as:</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{aligned}
\text{ARR} &amp;= \text{Risk in Control group} - \text{Risk in Treatment group} \\
&amp;= p_1 - p_2
\end{aligned}
\end{split}\]</div>
<p>And we can write AR as:</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{aligned}
\text{AR} &amp;= \text{Risk in Exposed group} - \text{Risk in Unexposed group} \\
&amp;= p_1 - p_2
\end{aligned}
\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(p_1\)</span> is the risk in the control or exposed group, and <span class="math notranslate nohighlight">\(p_2\)</span> the risk in the treatment or unexposed group. In the context of treatment evaluation, the placebo group is usually considered the reference, as we are interested in the effect of the treatment relative to the placebo.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_arr</span><span class="p">(</span><span class="n">table</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates absolute risk reduction / attributable risk from a 2x2 contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table (np.ndarray): A 2x2 NumPy array representing the contingency table.</span>
<span class="sd">            Rows: Control or Exposed/Treatment or Unexposed, Columns: Disease/No Disease</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: The absolute risk reduction.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Check if table is 2x2</span>
    <span class="k">if</span> <span class="n">table</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input table must be a 2x2 NumPy array.&quot;</span><span class="p">)</span>

    <span class="c1"># Calculate risk in exposed and unexposed groups</span>
    <span class="n">risk_control</span> <span class="o">=</span>   <span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:])</span>
    <span class="n">risk_treatment</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">table</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>

    <span class="c1"># Calculate attributable risk</span>
    <span class="n">ar</span> <span class="o">=</span> <span class="n">risk_control</span> <span class="o">-</span> <span class="n">risk_treatment</span>
    <span class="k">return</span> <span class="n">ar</span>

<span class="c1"># Application to our Apixaban example</span>
<span class="n">arr</span> <span class="o">=</span> <span class="n">calculate_arr</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Abosulte risk redution (ARR): </span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">arr</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Abosulte risk redution (ARR): 7.14%
</pre></div>
</div>
</div>
</div>
<p>The interpretation of ARR (or AR) depends on whether it’s positive or negative. A positive value indicates risk reduction, while a negative value indicates risk increase:</p>
<ul class="simple">
<li><p>ARR &gt; 0: this is the most desirable outcome. It indicates that the placebo group experiences a higher risk of the outcome than the treatment group. A positive ARR suggests a potential beneficial effect of the treatment, as it reduces the absolute risk.</p></li>
<li><p>ARR = 0: this suggests there’s no difference in risk between the placebo and treatment groups. The treatment, in this case, does not appear to alter the likelihood of the outcome.</p></li>
<li><p>ARR &lt; 0: this scenario is less common when evaluating treatment efficacy. It would imply that the placebo group has a lower risk of the outcome than the treatment group. While mathematically possible, this finding would raise concerns about potential biases or confounding factors in the study design or execution. It would warrant further investigation to understand why the placebo appears to be “protective” compared to the active treatment.</p></li>
</ul>
</section>
<section id="population-attributable-risk-par">
<h4>Population attributable risk (PAR)<a class="headerlink" href="#population-attributable-risk-par" title="Link to this heading">#</a></h4>
<p>While AR tells us how much extra risk an exposed group faces compared to an unexposed group, public health officials often want a broader perspective. They need to know how much of a disease’s overall burden in the <em>entire population</em> can be linked to a specific risk factor. This is where population attributable risk (PAR) comes in.</p>
<p>PAR quantifies the proportion of cases in the entire population that we could potentially prevent if we eliminated the risk factor. It gives us a powerful tool to assess the public health impact of an exposure:</p>
<div class="math notranslate nohighlight">
\[\text{PAR} = p_0 - p_2 = (p_1 - p_2) \times n_1 / N\]</div>
<p>where <span class="math notranslate nohighlight">\(p_0\)</span> is the overall proportion of the disease in the entire population, <span class="math notranslate nohighlight">\(p_1\)</span> the proportion of the disease in the exposed group, <span class="math notranslate nohighlight">\(p_2\)</span> the proportion of the disease in the unexposed group, <span class="math notranslate nohighlight">\(n_1\)</span> the number of exposed individuals and <span class="math notranslate nohighlight">\(N\)</span> the total number of individuals in the population, therefore <span class="math notranslate nohighlight">\(n_1 / N\)</span> is the <em>exposure prevalence</em>.</p>
<p>PAR isn’t just about how much riskier a factor makes things for exposed individuals (that’s AR). It also considers how common that exposure is in the population. The more widespread the exposure, the larger its potential impact on the entire population, even if the individual risk increase (AR) is relatively small.</p>
<p>For example, imagine a study where the contingency table looks like this:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Oral contraceptive</p></th>
<th class="head"><p>Thromboembolism</p></th>
<th class="head"><p>No thromboembolism</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>150</p></td>
<td><p>850</p></td>
</tr>
<tr class="row-odd"><td><p>No</p></td>
<td><p>50</p></td>
<td><p>950</p></td>
</tr>
</tbody>
</table>
</div>
<p>And let’s assume that 20% of the population uses oral contraceptives (the exposure in this case).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p_exposed</span>   <span class="o">=</span> <span class="mi">150</span> <span class="o">/</span> <span class="p">(</span><span class="mi">150</span> <span class="o">+</span> <span class="mi">850</span><span class="p">)</span>
<span class="n">p_unexposed</span> <span class="o">=</span>  <span class="mi">50</span> <span class="o">/</span> <span class="p">(</span> <span class="mi">50</span> <span class="o">+</span> <span class="mi">950</span><span class="p">)</span>
<span class="n">exposure_prevalence</span> <span class="o">=</span> <span class="mf">0.20</span>

<span class="n">PAR</span> <span class="o">=</span> <span class="p">(</span><span class="n">p_exposed</span> <span class="o">-</span> <span class="n">p_unexposed</span><span class="p">)</span> <span class="o">*</span> <span class="n">exposure_prevalence</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Population attributable risk (PAR): </span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">PAR</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Population attributable risk (PAR): 2.00%
</pre></div>
</div>
</div>
</div>
<p>The PAR of 0.02 (or 2%) indicates that if oral contraceptive use were eliminated from the population, we could potentially prevent 2% of all thromboembolism cases.</p>
</section>
<section id="number-needed-to-treat-ntt">
<h4>Number needed to treat (NTT)<a class="headerlink" href="#number-needed-to-treat-ntt" title="Link to this heading">#</a></h4>
<p>The number needed to treat (NNT), or number needed to harm (NNH), is a practical and clinically relevant measure that tells us how many patients need to receive a treatment in order to prevent <em>one additional adverse outcome</em> (or achieve one positive outcome). It’s calculated as the reciprocal of the absolute risk reduction:</p>
<div class="math notranslate nohighlight">
\[\text{NTT} = 1 / \text{ARR}\]</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_nnt</span><span class="p">(</span><span class="n">absolute_risk_reduction</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates number needed to treat (NNT) or number needed to harm (NNH).</span>

<span class="sd">    Args:</span>
<span class="sd">        absolute_risk_reduction (float): The absolute risk reduction (risk difference or AR).</span>

<span class="sd">    Returns:</span>
<span class="sd">        int or float: </span>
<span class="sd">            - If ARR &gt; 0: NNT (rounded to nearest integer).</span>
<span class="sd">            - If ARR &lt; 0: NNH (rounded to nearest integer).</span>
<span class="sd">            - If ARR = 0: None (no effect)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">absolute_risk_reduction</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">round</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">absolute_risk_reduction</span><span class="p">)</span>  <span class="c1"># NNT</span>
    <span class="k">elif</span> <span class="n">absolute_risk_reduction</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">round</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span> <span class="n">absolute_risk_reduction</span><span class="p">)</span> <span class="c1"># NNH</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>  <span class="c1"># No effect</span>

<span class="c1"># Application to our Apixaban example</span>
<span class="n">nnt</span> <span class="o">=</span> <span class="n">calculate_nnt</span><span class="p">(</span><span class="n">arr</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number needed to treat (NNT): </span><span class="si">{</span><span class="n">nnt</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number needed to treat (NNT): 14
</pre></div>
</div>
</div>
</div>
<p>A <em>lower NNT</em> indicates a <em>more effective</em> treatment. For example, an NNT of 5 means that we need to treat only 5 patients to prevent one additional bad outcome compared to the control group. A <em>higher NNT</em> indicates a <em>less effective</em> treatment. For example, an NNT of 100 means we need to treat 100 patients to prevent one additional bad outcome.</p>
</section>
<section id="relative-risk-rr">
<h4>Relative risk (RR)<a class="headerlink" href="#relative-risk-rr" title="Link to this heading">#</a></h4>
<p>The relative risk (RR), also known as the <strong>risk ratio</strong>, compares the risk of an event (such as disease recurrence) occurring in one group versus another. It’s a powerful tool for quantifying how much more (or less) likely an outcome is when exposed to a particular factor or treatment.</p>
<p>In its simplest form, the RR is calculated by dividing the risk in the exposed group by the risk in the unexposed (or control) group:</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{aligned}
\text{RR} &amp;= \frac{\text{Risk in exposed group}}{\text{Risk in unexposed group}} \\
&amp;= \frac{p_1}{p_2} \\
&amp;= \frac{a / (a + b)}{c / (c + d)}
\end{aligned}
\end{split}\]</div>
<p>when the contingency table looks like this:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p>Event</p></th>
<th class="head"><p>No event</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Exposed</p></td>
<td><p>a</p></td>
<td><p>b</p></td>
</tr>
<tr class="row-odd"><td><p>Unexposed</p></td>
<td><p>c</p></td>
<td><p>d</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_rr</span><span class="p">(</span><span class="n">table_or_p1</span><span class="p">,</span> <span class="n">p2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates relative risk (RR) from a contingency table or proportions.</span>

<span class="sd">    Args:</span>
<span class="sd">        table_or_p1: Either a 2x2 NumPy array representing a contingency table, </span>
<span class="sd">        or the proportion (float) of events in the exposed group.</span>
<span class="sd">        p2 (float, optional): If table_or_p1 is a proportion, this should be </span>
<span class="sd">        the proportion of events in the unexposed group.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: The relative risk (RR) or None if calculation is not possible.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">p2</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Input is a contingency table</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">table_or_p1</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">or</span> <span class="n">table_or_p1</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid input: Must provide a 2x2 NumPy array.&quot;</span><span class="p">)</span>

        <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_or_p1</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span>
        <span class="n">p2</span> <span class="o">=</span> <span class="n">c</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Input is already proportions</span>
        <span class="n">p1</span> <span class="o">=</span> <span class="n">table_or_p1</span>

    <span class="c1"># Check for division by zero</span>
    <span class="k">if</span> <span class="n">p2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">rr</span> <span class="o">=</span> <span class="n">p1</span> <span class="o">/</span> <span class="n">p2</span>
    <span class="k">return</span> <span class="n">rr</span>


<span class="c1"># Calculate from contingency table</span>
<span class="n">rr</span> <span class="o">=</span> <span class="n">calculate_rr</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Relative risk (from table): </span><span class="si">{</span><span class="n">rr</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Calculate from proportions</span>
<span class="n">p2</span> <span class="o">=</span> <span class="mi">73</span> <span class="o">/</span> <span class="mi">829</span>  <span class="c1"># placebo</span>
<span class="n">p1</span> <span class="o">=</span> <span class="mi">14</span> <span class="o">/</span> <span class="mi">840</span>  <span class="c1"># Apixaban</span>

<span class="n">rr_from_proportions</span> <span class="o">=</span> <span class="n">calculate_rr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Relative risk (from proportions, with reference = placebo): </span><span class="si">{</span><span class="n">rr_from_proportions</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>


<span class="c1"># Alternative method for obtaining the RR using scipy.stats.contingency.relative_risk</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Relative risk (from scipy.stats.contingency): </span><span class="si">{</span><span class="n">contingency</span><span class="o">.</span><span class="n">relative_risk</span><span class="p">(</span>
<span class="w">    </span><span class="n">exposed_cases</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
<span class="w">    </span><span class="n">exposed_total</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
<span class="w">    </span><span class="n">control_cases</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
<span class="w">    </span><span class="n">control_total</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span>
<span class="p">)</span><span class="o">.</span><span class="n">relative_risk</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Relative risk (from table): 5.283
Relative risk (from proportions, with reference = placebo): 0.189
Relative risk (from scipy.stats.contingency): 5.283
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p>RR &gt; 1: the risk is higher in the ‘exposed’ group, <em>i.e., the first row of the contingency table</em>). For example, an RR of 2 means the exposed group has twice the risk of the event compared to the unexposed group.</p></li>
<li><p>RR = 1: the risk is the same in both groups (no association).</p></li>
<li><p>RR &lt; 1: the risk is lower in the ‘exposed’ group. In this case, the exposure is considered a protective factor. For instance, an RR of 0.5 means the exposed group has half the risk of the event compared to the unexposed group.</p></li>
</ul>
<p><u>Note:</u> The order of the rows in the contingency table matters for interpreting the relative risk (RR). If the first row represents the placebo group and the second row represents the Apixaban group, the fundamental meaning of the RR value itself doesn’t change, but the reference group against which we’re comparing does.</p>
<p>When we calculated the RR using our original contingency table, with the <em>placebo group as the reference in the first row</em>, we obtained an RR of 5.283. This means that subjects in the placebo group were over <em>five times more likely</em> to experience a recurrent disease compared to those who received the treatment.</p>
<p>Alternatively, if we directly submitted the proportions <span class="math notranslate nohighlight">\(p_1\)</span> (risk in the treatment group) and <span class="math notranslate nohighlight">\(p_2\)</span> (risk in the placebo group) to the <code class="docutils literal notranslate"><span class="pre">calculate_rr</span></code> function, we obtain an RR of 0.189. This can be interpreted in two ways:</p>
<ol class="arabic simple">
<li><p>Risk reduction: patients receiving the drug had only 18.9% of the risk of recurrent disease compared to those receiving the placebo. In other words, the treatment significantly reduced the risk.</p></li>
<li><p>Protective effect: the treatment exhibited a protective effect, reducing the relative risk of recurrent disease by a substantial margin of 1 - 0.189 = 81.1%.</p></li>
</ol>
<p>The reciprocal of the RR, i.e., <span class="math notranslate nohighlight">\(1/\text{RR}\)</span>, provides the inverse interpretation. For example, if <span class="math notranslate nohighlight">\(\text{RR} = 0.2\)</span>, then <span class="math notranslate nohighlight">\(1/\text{RR} = 5\)</span>, meaning the first group has 5 times the risk of the second group.</p>
</section>
<section id="logrr">
<h4>LogRR<a class="headerlink" href="#logrr" title="Link to this heading">#</a></h4>
<p>The log transformation of the relative risk log(RR) is a valuable tool in biostatistical analysis:</p>
<ul class="simple">
<li><p>Symmetry: RR is not symmetric around 1. An RR of 2 (doubling of risk) is not the opposite of an RR of 0.5 (halving of risk). However, the log transformation makes the scale symmetric around 0. This means that a log(RR) of 0.693 (corresponding to an RR of 2) is the exact opposite of a log(RR) of -0.693 (corresponding to an RR of 0.5). This symmetry is helpful for statistical calculations and interpretations.</p></li>
<li><p>Normal approximation: the distribution of log(RR) tends to be more normally distributed than the distribution of RR itself. Many statistical tests assume a normal distribution, so working with log(RR) allows us to use these tests more appropriately.</p></li>
<li><p>Confidence intervals: when calculating confidence intervals for RR, a log transformation can make the intervals more symmetrical and accurate, especially when the RR is far from 1.</p></li>
<li><p>Meta-analysis: in meta-analyses that combine results from multiple studies, log(RR) is often used as the effect measure. This is because log(RR) allows for easier pooling of results and better statistical properties.</p></li>
</ul>
<p>Calculating the log(RR) is straightforward. You simply take the natural logarithm of the relative risk:</p>
<div class="math notranslate nohighlight">
\[\log(\text{RR}) = \ln(\text{RR})\]</div>
<p>As we saw, the RR is the ratio of the risk in the exposed group <span class="math notranslate nohighlight">\(p_1\)</span> to the risk in the unexposed group <span class="math notranslate nohighlight">\(p_2\)</span>, so we can also write:</p>
<div class="math notranslate nohighlight">
\[\log(\text{RR}) = \log\left(\frac{p_1}{p_2}\right) = \log{p_1} - \log{p_2}\]</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">math</span>

<span class="n">log_rr</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">rr</span><span class="p">)</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Log(RR): </span><span class="si">{</span><span class="n">log_rr</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Corresponding relative risk: </span><span class="si">{</span><span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">log_rr</span><span class="p">)</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Log(RR): 1.665
Corresponding relative risk: 5.283
</pre></div>
</div>
</div>
</div>
</section>
<section id="odds-ratio-or">
<h4>Odds ratio (OR)<a class="headerlink" href="#odds-ratio-or" title="Link to this heading">#</a></h4>
<p>The odds ratio (OR) is another way to quantify the association between an exposure and an outcome. While relative risk compares the probabilities of an event occurring in two groups, the odds ratio compares the <strong>odds</strong> of the event.</p>
<p>Odds are slightly different from probabilities. The odds of an event are calculated as the probability of the event happening divided by the probability of it not happening:</p>
<div class="math notranslate nohighlight">
\[\text{Odds} = \frac{\text{Probability of event}}{\text{Probability of no event}} = \frac{p}{1-p}\]</div>
<p>OR is then the ratio of the odds of the event in the exposed group to the odds of the event in the unexposed group:</p>
<div class="math notranslate nohighlight">
\[\text{OR} = \frac{\text{Odds in exposed group}}{\text{Odds in unexposed group}}\]</div>
<p>Mathematically, the OR formula is:</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{aligned}
\text{OR} &amp;= \frac{p_1 / (1 - p_1)}{p_2 / (1 - p_2)} \\
&amp;= \frac{a / (a + b)}{c / (c + d)} \times \frac{1 - c / (c + d)}{1 - a / (a + b)} \\
&amp;= \frac{a / (a + b)}{c / (c + d)} \times \frac{\frac{\cancel{c} + d - \cancel{c}}{c + d}}{\frac{\cancel{a} + b - \cancel{a}}{a + b}} \\
&amp;= \frac{a \cancel{(c + d)}}{c \cancel{(a + b)}} \times \frac{d \cancel{(a + b)}}{b \cancel{(c + d)}} \\
&amp;= \frac{ad}{bc}
\end{aligned}
\end{split}\]</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_or</span><span class="p">(</span><span class="n">table</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates odds ratio (OR) from a 2x2 contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table: A 2x2 NumPy array representing the contingency table</span>
<span class="sd">               in the format [[a, b], [c, d]].</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: The odds ratio (OR).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">or</span> <span class="n">table</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid input: Must provide a 2x2 NumPy array.&quot;</span><span class="p">)</span>

    <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">a</span> <span class="o">*</span> <span class="n">d</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">b</span> <span class="o">*</span> <span class="n">c</span><span class="p">)</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">oddsratio</span> <span class="o">=</span> <span class="n">calculate_or</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Odds ratio: </span><span class="si">{</span><span class="n">oddsratio</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;(Relative risk was: </span><span class="si">{</span><span class="n">rr</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

<span class="c1"># Alternative method for obtaining the RR using scipy.stats.contingency.odds_ratio</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Odds ratio (from scipy.stats.contingency): </span><span class="si">{</span><span class="n">contingency</span><span class="o">.</span><span class="n">odds_ratio</span><span class="p">(</span>
<span class="w">        </span><span class="n">table</span><span class="o">=</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span>
<span class="w">        </span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span>
<span class="w">    </span><span class="p">)</span><span class="o">.</span><span class="n">statistic</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Odds ratio: 5.697
(Relative risk was: 5.283)
Odds ratio (from scipy.stats.contingency): 5.697
</pre></div>
</div>
</div>
</div>
<p>The odds ratio and relative risk are related but not identical.  The key distinction is that the RR deals with probabilities, while the OR deals with odds:</p>
<ul class="simple">
<li><p>OR &gt; 1: the odds of the event are higher in the ‘exposed’/reference group, i.e., in the current case the placebo/control group.</p></li>
<li><p>OR = 1: the odds of the event are the same in both groups (no association).</p></li>
<li><p>OR &lt; 1: the odds of the event are lower in the ‘exposed’/reference group, suggesting a protective effect of the exposure.</p></li>
</ul>
<p>When the event is rare (low incidence or prevalence), the <em>OR and RR are very close</em> in value and can be interpreted similarly. In contrast, when the event is more common, the OR tends to overestimate the magnitude of association compared to the RR. The more common the event, the greater the discrepancy.</p>
<p>Actually, <a class="reference external" href="https://docs.scipy.org/doc/scipy-1.13.1/reference/generated/scipy.stats.fisher_exact.html"><code class="docutils literal notranslate"><span class="pre">scipy.stats.fisher_exact</span></code></a> provides the odds ratio directly from the contingency table, as seen in the section about P value.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Odds ratio (from `fisher_exact`): </span><span class="si">{</span><span class="n">odds_ratio</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Odds ratio (from `fisher_exact`): 5.697
</pre></div>
</div>
</div>
</div>
</section>
<section id="logor">
<h4>LogOR<a class="headerlink" href="#logor" title="Link to this heading">#</a></h4>
<p>Similar to log(RR), the log(OR) transformation creates symmetry around zero, where log(OR) = 0 indicates no effect (OR = 1). Positive log(OR) values indicate increased odds, while negative values indicate decreased odds. It can be interpreted as the <em>additive effect</em> of a predictor on the log odds of the outcome. This can be helpful in understanding the relative contributions of multiple predictors to the outcome.</p>
<p>Additionally, log(OR) tends to be more normally distributed than the OR itself, making it suitable for many statistical tests that assume normality.</p>
<p>Calculating confidence intervals (CIs) for the OR directly can result in asymmetrical and potentially inaccurate intervals, especially when the OR is far from 1. By taking the log(OR), we can calculate CIs on the log scale, which tend to be more symmetric and reliable. We can then transform these CIs back to the original OR scale to get meaningful estimates of the association’s precision.</p>
<p>Moreover, log(OR) is the preferred effect measure in many <em>meta-analyses</em>. This allows for easier pooling of results from different studies and facilitates statistical calculations like weighting studies based on their precision.</p>
<p>Finally, in logistic regression, a common statistical model for analyzing binary outcomes, the log(OR) is directly estimated as the coefficient for a predictor variable. This makes log(OR) a fundamental building block for understanding the relationships between predictors and outcomes in these models.</p>
<p>Because <span class="math notranslate nohighlight">\(\text{OR} = \frac{p_1/(1 - p_1)}{p_2/(1 - p_2)}\)</span>, we can write:</p>
<div class="math notranslate nohighlight">
\[\log(\text{OR}) = \ln(\text{OR}) = \log(p_1) - \log(1 - p_1) - \log(p_2) + \log(1 - p_2)\]</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">log_odds_ratio</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">odds_ratio</span><span class="p">)</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Log(OR) as the natural logarithm of OR: </span><span class="si">{</span><span class="n">log_odds_ratio</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># log(OR) obtained with calculations</span>
<span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">p1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span>
<span class="n">p2</span> <span class="o">=</span> <span class="n">c</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>

<span class="n">log_odds_ratio_calc</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p1</span><span class="p">)</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p1</span><span class="p">)</span> <span class="o">-</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Log(OR) obtained with the manual calculation: </span><span class="si">{</span><span class="n">log_odds_ratio_calc</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Odds ratio obtained for the log(OR)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Odds ratio calculated as the exponential of log(OR): </span><span class="si">{</span><span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">log_odds_ratio_calc</span><span class="p">)</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Log(OR) as the natural logarithm of OR: 1.740
Log(OR) obtained with the manual calculation: 1.740
Odds ratio calculated as the exponential of log(OR): 5.697
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="confidence-intervals">
<h3>Confidence intervals<a class="headerlink" href="#confidence-intervals" title="Link to this heading">#</a></h3>
<p>In the previous sections, we’ve explored several key measures for comparing proportions: ARR, NNT, RR, and OR. However, these values are just point estimates derived from our sample data. To truly understand the reliability of these estimates and draw meaningful conclusions about the broader population, we need to consider their confidence intervals (CIs).</p>
<p>A confidence interval provides a <em>range of plausible values for a population parameter</em> (like ARR, NTT, RR, or OR) based on the sample data. It tells us how confident we can be that the true population value falls within that range. A 95% confidence interval, for example, means that if we were to repeat our study many times, we would expect <em>95% of the calculated intervals to contain the true population value</em>.</p>
<p>Confidence intervals are crucial for several reasons:</p>
<ul class="simple">
<li><p>Estimating precision: they quantify the uncertainty surrounding our point estimates. A wider CI indicates more uncertainty, while a narrower CI suggests greater <strong>precision</strong>.</p></li>
<li><p>Statistical significance: if a CI for a measure like RR or OR does not include 1, it typically indicates a <em>statistically significant association</em> between the exposure and outcome.</p></li>
<li><p>Clinical relevance: CIs help us assess the practical importance of our findings. A wide CI around a large RR might suggest that the effect could be clinically meaningful, even if the exact magnitude is uncertain.</p></li>
</ul>
<p>The methods for calculating confidence intervals vary depending on the specific measure (ARR, NNT, RR, OR) and the underlying statistical assumptions, e.g., normal approximation (a simple approach that works for large samples), exact methods (more accurate for smaller samples), and bootstrapping (a versatile resampling technique that can be applied to various scenarios), though they are typically calculated as:</p>
<div class="math notranslate nohighlight">
\[\text{CI}^{1 - \alpha} = \hat{\theta} \pm c^\ast \times s_{\hat{\theta}}\]</div>
<p>where <span class="math notranslate nohighlight">\(\text{CI}^{1 - \alpha}\)</span> represents the confidence interval at <span class="math notranslate nohighlight">\(1 - \alpha\)</span> confidence level, <span class="math notranslate nohighlight">\(\hat{\theta}\)</span> the point estimate of the population parameter, <span class="math notranslate nohighlight">\(c^\ast\)</span> the critical value from the relevant distribution corresponding to the desired confidence level and if relevant, given degrees of freedom, e.g., t-distribution (<span class="math notranslate nohighlight">\(t_{\alpha/2, df}\)</span>), standard normal (<span class="math notranslate nohighlight">\(z_{\alpha/2}\)</span>), etc., and <span class="math notranslate nohighlight">\(s_{\hat{\theta}}\)</span> the standard error (SE) of the point estimate.</p>
<section id="ci-for-proportions">
<h4>CI for proportions<a class="headerlink" href="#ci-for-proportions" title="Link to this heading">#</a></h4>
<p>As discussed in our earlier chapter on confidence intervals for proportions, several methods can be employed to estimate the CI around the proportion of patients experiencing disease progression. In this analysis, we’ll focus on the proportion of patients with disease progression in both the placebo group (73 out of 829) and the apixaban-treated group (14 out of 840). Let’s utilize the ‘normal’ method available in the <code class="docutils literal notranslate"><span class="pre">statsmodels.stats.proportion</span></code> module for this calculation.</p>
<p>The normal approximation is considered valid when the sample size is sufficiently large. A common rule of thumb is that both of the following conditions should hold: <span class="math notranslate nohighlight">\(n \times p \ge 5\)</span> and <span class="math notranslate nohighlight">\(n \times (1 - p) \ge 5\)</span>. <em>The underlying true proportion is not too close to 0 or 1. If the proportion is very close to either extreme, other methods like the <a class="reference external" href="https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval">Wilson score interval</a> might be more appropriate.</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.proportion</span> <span class="kn">import</span> <span class="n">proportion_confint</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span> <span class="c1"># statsmodels gives a lot of warnings</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

<span class="c1"># Set 95% confidence interval</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span>

<span class="c1"># Calculate CI for placebo group (using normal approximation)</span>
<span class="n">ci_placebo</span> <span class="o">=</span> <span class="n">proportion_confint</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">nobs</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">conf</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">)</span>

<span class="c1"># Calculate CI for apixaban group</span>
<span class="n">ci_apixaban</span> <span class="o">=</span> <span class="n">proportion_confint</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="n">c</span><span class="p">,</span> <span class="n">nobs</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">conf</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Proportion and </span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">conf</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">% CI&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="o">*</span><span class="mi">21</span><span class="p">)</span>
<span class="c1"># We round each bound of the returned tuples using a generator</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;placebo group:</span><span class="se">\t</span><span class="si">{</span><span class="n">a</span><span class="o">/</span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">)</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bound</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">bound</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">ci_placebo</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;apixaban group:</span><span class="se">\t</span><span class="si">{</span><span class="n">c</span><span class="o">/</span><span class="p">(</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">)</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bound</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">bound</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">ci_apixaban</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Proportion and 95% CI
---------------------
placebo group:	0.0881, (0.069, 0.107)
apixaban group:	0.0167, (0.008, 0.025)
</pre></div>
</div>
</div>
</div>
</section>
<section id="ci-for-arr">
<h4>CI for ARR<a class="headerlink" href="#ci-for-arr" title="Link to this heading">#</a></h4>
<section id="standard-approximations">
<h5>Standard approximations<a class="headerlink" href="#standard-approximations" title="Link to this heading">#</a></h5>
<p>While the ARR gives us a single point estimate of the treatment effect, the confidence interval provides a range of plausible values within which we can be 95% confident that the true population ARR lies.</p>
<p>The confidence interval for ARR is calculated as follows:</p>
<div class="math notranslate nohighlight">
\[\text{CI} = \text{ARR} \pm z^\ast \times s_\text{ARR}\]</div>
<p>where <span class="math notranslate nohighlight">\(\text{ARR} = p_1 - p_2\)</span> is the absolute risk reduction, <span class="math notranslate nohighlight">\(z^\ast\)</span> is the critical z-value from the standard normal distribution corresponding to the desired level of confidence (e.g., 1.96 for a 95% CI), and <span class="math notranslate nohighlight">\(s_\text{ARR} = s_{p_1 - p_2}\)</span> is the standard error of the absolute risk reduction.</p>
<p>Because the standard error of a difference between two proportions is calculated as <span class="math notranslate nohighlight">\(s_{p_1 - p_2} = \sqrt{\mathrm{Var}(p_1) + \mathrm{Var}(p_2)}\)</span>, the variance of a sample proportion <span class="math notranslate nohighlight">\(x\)</span> is estimated as <span class="math notranslate nohighlight">\(\mathrm{Var}(x) = x (1 - x)/n\)</span> where <span class="math notranslate nohighlight">\(n\)</span> is the sample size, and we can express <span class="math notranslate nohighlight">\(p_1\)</span> and <span class="math notranslate nohighlight">\(p_2\)</span> in terms of the ovarall risk <span class="math notranslate nohighlight">\(p\)</span> and the ARR, substituting these into the formula for <span class="math notranslate nohighlight">\(s_{p_1 - p_2}\)</span> and simplifying, we arrive at:</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{align}
s_{p_1 - p_2} &amp;= \sqrt{p_1 (1 - p_1) /n_1 + p_2 (1 - p_2) /n_2} \\
&amp;\approx \sqrt{p (1 - p) /n_1 + p (1 - p) /n_2} \\
&amp;\approx \sqrt{p (1 - p) (1/n_1 + 1/n_2)}
\end{align}
\end{split}\]</div>
<p>by approximating both <span class="math notranslate nohighlight">\(p_1\)</span> and <span class="math notranslate nohighlight">\(p_2\)</span> using the overall risk <span class="math notranslate nohighlight">\(p\)</span>, calculated as <span class="math notranslate nohighlight">\(p = m_1/N\)</span>, where <span class="math notranslate nohighlight">\(m_1\)</span> is the total number of events in the population <span class="math notranslate nohighlight">\((a + c)\)</span>, and <span class="math notranslate nohighlight">\(N\)</span> is the total sample size <span class="math notranslate nohighlight">\((a + b + c + d)\)</span>, and <span class="math notranslate nohighlight">\(n_1\)</span> and <span class="math notranslate nohighlight">\(n_2\)</span> are the sample sizes of the control/exposed and treatment/unexposed groups respectively, because we are dealing with <em>rare events</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">norm</span>

<span class="k">def</span> <span class="nf">calculate_arr_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates absolute risk reduction (ARR) and its confidence interval (CI) </span>
<span class="sd">    from a 2x2 contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table (np.ndarray): A 2x2 NumPy array representing the contingency table.</span>
<span class="sd">        conf (float, optional): Confidence level (default: 0.95).</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing:</span>
<span class="sd">            - ARR (float): The absolute risk reduction.</span>
<span class="sd">            - ci_lower (float): Lower bound of the confidence interval.</span>
<span class="sd">            - ci_upper (float): Upper bound of the confidence interval.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">table</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input table must be a 2x2 array.&quot;</span><span class="p">)</span>

    <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

    <span class="c1"># Risk calculations</span>
    <span class="n">p1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span>
    <span class="n">p2</span> <span class="o">=</span> <span class="n">c</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">p1</span> <span class="o">-</span> <span class="n">p2</span>

    <span class="c1"># Standard error of AR</span>
    <span class="n">N</span> <span class="o">=</span> <span class="n">a</span> <span class="o">+</span> <span class="n">b</span> <span class="o">+</span> <span class="n">c</span> <span class="o">+</span> <span class="n">d</span>  <span class="c1"># total number of observations</span>
    <span class="n">p_bar</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">c</span><span class="p">)</span> <span class="o">/</span> <span class="n">N</span> <span class="c1"># overall risk</span>
    <span class="n">se_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">p_bar</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p_bar</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)))</span>

    <span class="c1"># Z-score for desired confidence level</span>
    <span class="n">z</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">ppf</span><span class="p">((</span><span class="mi">1</span> <span class="o">+</span> <span class="n">conf</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Confidence interval (adjust for negative AR if necessary)</span>
    <span class="n">ci_lower</span> <span class="o">=</span> <span class="n">arr</span> <span class="o">-</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_arr</span>
    <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">arr</span> <span class="o">+</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_arr</span>
   
    <span class="k">return</span> <span class="n">arr</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span> <span class="c1"># Set 95% confidence interval</span>
<span class="n">arr</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">calculate_arr_ci</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="n">conf</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Absolute risk reduction (ARR): </span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">arr</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">conf</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">% confidence interval for ARR: (</span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">ci_lower</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%, </span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">ci_upper</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Absolute risk reduction (ARR): 7.139%
95% confidence interval for ARR: (5.01%, 9.27%)
</pre></div>
</div>
</div>
</div>
<p>This analysis shows a 95% confidence interval for the absolute risk reduction associated with apixaban treatment ranging from 5.01% to 9.27%. If we assume our study participants are representative of the broader population of adults with thromboembolism, we can be 95% confident that apixaban treatment will reduce the risk of disease progression within this population by a value somewhere between 5.01% and 9.27%.</p>
</section>
<section id="z-test-for-comparing-proportions">
<h5>Z-test for comparing proportions<a class="headerlink" href="#z-test-for-comparing-proportions" title="Link to this heading">#</a></h5>
<p>When dealing with large sample sizes and comparing proportions between two groups, the z-test offers a <em>simpler alternative</em> to Fisher’s exact test. This test assesses whether there’s a statistically significant difference between two proportions and provides insights into the magnitude of the effect.</p>
<p>It assumes that the sampling distribution of the difference in proportions is <em>approximately normal</em>, which holds true for large sample sizes (typically at least 30 in group). The z-statistic can be writent as:</p>
<div class="math notranslate nohighlight">
\[z = \frac{\text{ARR}}{s_\text{ARR}} = \frac{p_1 - p_2}{s_{p_1 - p_2}}\]</div>
<p>where <span class="math notranslate nohighlight">\(p_1\)</span> is the proportion in group 1, <span class="math notranslate nohighlight">\(p_2\)</span> the proportion in group 2, and <span class="math notranslate nohighlight">\(s_{p_1 - p_2}\)</span> the standard error of the difference in proportions.</p>
<p>Notice that the numerator in the z-statistic is precisely the ARR we are interested in. The denominator is the same standard error we used to calculate the CI for ARR. This highlights the direct link between the z-test and the confidence interval.</p>
<p>We can implement the z-test for comparing proportions in Python using the <a class="reference external" href="https://www.statsmodels.org/stable/generated/statsmodels.stats.proportion.proportions_ztest.html"><code class="docutils literal notranslate"><span class="pre">proportions_ztest</span></code> function from <code class="docutils literal notranslate"><span class="pre">statsmodels.stats.proportion</span></code></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.proportion</span> <span class="kn">import</span> <span class="n">proportions_ztest</span>

<span class="c1"># Extract counts and total observations for each group from the Apixaban table</span>
<span class="n">count</span> <span class="o">=</span> <span class="n">table_df</span><span class="p">[</span><span class="s1">&#39;recurrence&#39;</span><span class="p">]</span>
<span class="n">nobs</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Perform Z-test for difference in proportions</span>
<span class="n">z_stat</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">proportions_ztest</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">nobs</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Z-statistic:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">z_stat</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;P value:&quot;</span><span class="p">,</span> <span class="n">p_value</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Z-statistic: 6.56
P value: 5.368112856740949e-11
</pre></div>
</div>
</div>
</div>
<p>The z-test assesses whether the observed ARR is significantly different from zero (no difference in risk). If the absolute value of the z-statistic is large enough (typically exceeding 1.96 for a 95% confidence level), the P value associated with the Z-test will be small, indicating statistical significance. A statistically significant result in the Z-test implies that the <em>CI for ARR will not include zero</em>. This means we can be confident that the true ARR in the population is not zero, and there is a real difference in risk between the two groups.</p>
</section>
<section id="exact-methods">
<h5>Exact methods<a class="headerlink" href="#exact-methods" title="Link to this heading">#</a></h5>
<p>Using the overall risk <span class="math notranslate nohighlight">\(p\)</span> as an approximation for both <span class="math notranslate nohighlight">\(p_1\)</span> and <span class="math notranslate nohighlight">\(p_2\)</span> in the <span class="math notranslate nohighlight">\(s_{\text{ARR}}\)</span> formula introduces only a small error. This approximation simplifies the calculation without sacrificing much accuracy when dealing with rare events. This approximation is less accurate when the event is not rare, and/or the sample size is small. For common events, it’s preferable to use the <strong>exact formula</strong> with the individual risk estimates (<span class="math notranslate nohighlight">\(p_1\)</span> and <span class="math notranslate nohighlight">\(p_2\)</span>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Application with the Apixaban example</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">n1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">+</span> <span class="n">b</span>
<span class="n">n2</span> <span class="o">=</span> <span class="n">c</span> <span class="o">+</span> <span class="n">d</span>

<span class="c1"># True proportions</span>
<span class="n">p1_true</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="n">n1</span>
<span class="n">p2_true</span> <span class="o">=</span> <span class="n">c</span> <span class="o">/</span> <span class="n">n2</span>

<span class="c1"># Overall risk</span>
<span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">n1</span> <span class="o">*</span> <span class="n">p1_true</span> <span class="o">+</span> <span class="n">n2</span> <span class="o">*</span> <span class="n">p2_true</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">n1</span> <span class="o">+</span> <span class="n">n2</span><span class="p">)</span>
<span class="c1"># same as `p_bar` in the previous calculate_arr_ci function</span>

<span class="c1"># Calculate standard error using exact proportions</span>
<span class="n">se_exact</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">((</span><span class="n">p1_true</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p1_true</span><span class="p">)</span> <span class="o">/</span> <span class="n">n1</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">p2_true</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p2_true</span><span class="p">)</span> <span class="o">/</span> <span class="n">n2</span><span class="p">))</span>

<span class="c1"># Calculate standard error using overall risk (approximation)</span>
<span class="n">se_approx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">p</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">n1</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">n2</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Overall risk:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Exact SE:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">se_exact</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Approximated SE:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">se_approx</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Overall risk: 0.052
Exact SE: 0.01079
Approximated SE: 0.01088
</pre></div>
</div>
</div>
</div>
<p>Moreover, standard methods for calculating the CI for ARR, like the normal approximation can be inaccurate when:</p>
<ul class="simple">
<li><p>Sample sizes are small.</p></li>
<li><p>The observed proportions are close to 0 or 1.</p></li>
<li><p>We want a more precise estimate of the CI.</p></li>
</ul>
<p>The <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/9595617/">Newcombe/Wilson score method</a> often provides a more reliable and accurate CI. It is a hybrid approach that combines the strengths of two other methods for calculating CIs for risk differences:</p>
<ul class="simple">
<li><p>Wilson score interval: the Wilson score interval is a well-regarded method for calculating CIs for individual proportions (<span class="math notranslate nohighlight">\(p_1\)</span> and <span class="math notranslate nohighlight">\(p_2\)</span>). It provides good coverage probability and performs well even for small sample sizes or proportions close to 0 or 1.</p></li>
<li><p>Newcombe’s hybrid score interval: this method extends the Wilson score interval to the difference of two proportions (ARR). It calculates separate Wilson score intervals for <span class="math notranslate nohighlight">\(p_1\)</span> and <span class="math notranslate nohighlight">\(p_2\)</span>, and then combines them to obtain a CI for the ARR.</p></li>
</ul>
<p>The Wilson score interval for a proportion <span class="math notranslate nohighlight">\(p\)</span> is given by:</p>
<div class="math notranslate nohighlight">
\[\text{CI}(p) = \frac{p + z^2 / 2n \pm z \times \sqrt{p(1 - p) / n + z^2 / 4n^2}}{1 + z^2 / n}\]</div>
<p>where <span class="math notranslate nohighlight">\(p\)</span> is the sample proportion, <span class="math notranslate nohighlight">\(n\)</span> is the sample size, and <span class="math notranslate nohighlight">\(z\)</span> is the z-score corresponding to the desired confidence level (e.g., 1.96 for a 95% CI).</p>
<p>We can use the <a class="reference external" href="https://www.statsmodels.org/dev/generated/statsmodels.stats.proportion.confint_proportions_2indep.html"><code class="docutils literal notranslate"><span class="pre">statsmodels.stats.proportion.confint_proportions_2indep</span></code></a> function to calculate the Newcombe CI for the ARR by setting <code class="docutils literal notranslate"><span class="pre">compare='diff'</span></code> and <code class="docutils literal notranslate"><span class="pre">method='newcomb'</span></code> arguments. This function also offers other methods, including ‘wald’ (which is related to the Katz log method) and ‘score’ (which is related to the Koopman asymptotic score method).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.proportion</span> <span class="kn">import</span> <span class="n">confint_proportions_2indep</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

<span class="n">ci_arr_newcombe</span> <span class="o">=</span> <span class="n">confint_proportions_2indep</span><span class="p">(</span>
    <span class="n">count1</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">nobs1</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">count2</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">nobs2</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span>
    <span class="n">compare</span><span class="o">=</span><span class="s1">&#39;diff&#39;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;newcomb&#39;</span><span class="p">,</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;95% Newcombe CI for ARR: </span><span class="si">{</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bound</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">bound</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">ci_arr_newcombe</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>95% Newcombe CI for ARR: (0.051, 0.094)
</pre></div>
</div>
</div>
</div>
<p>While these methods provide approximations to the Koopman and Katz methods, they are often sufficient for many practical purposes. For a more precise calculation using the Koopman asymptotic score or Katz log methods, we may need to explore alternative implementations or specialized packages.</p>
</section>
</section>
<section id="ci-for-nnt">
<h4>CI for NNT<a class="headerlink" href="#ci-for-nnt" title="Link to this heading">#</a></h4>
<p>The confidence interval for the NNT provides a range of plausible values for the true NNT in the population. Since NNT is the reciprocal of ARR, the confidence interval for NNT will be wider when the CI for ARR is narrow, and vice versa. This reflects the fact that smaller treatment effects (smaller ARR) lead to larger NNT values with more uncertainty.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_nnt_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates NNT (or NNH) and its confidence interval from a 2x2 contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table (np.ndarray): A 2x2 NumPy array representing the contingency table.</span>
<span class="sd">        conf (float, optional): Confidence level (default: 0.95).</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing:</span>
<span class="sd">            - nnt_or_nnh (float): NNT if ARR &gt; 0, NNH if ARR &lt; 0, None if ARR = 0.</span>
<span class="sd">            - ci_lower (float or str): Lower bound of CI (or &quot;Infinity&quot; if undefined).</span>
<span class="sd">            - ci_upper (float or str): Upper bound of CI (or &quot;Infinity&quot; if undefined).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">arr</span><span class="p">,</span> <span class="n">ci_lower_arr</span><span class="p">,</span> <span class="n">ci_upper_arr</span> <span class="o">=</span> <span class="n">calculate_arr_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">arr</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Calculate NNT</span>
        <span class="n">nnt_or_nnh</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">arr</span><span class="p">)</span>
        <span class="n">ci_lower</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">ci_upper_arr</span><span class="p">)</span>  <span class="c1"># Invert CI for NNT</span>
        <span class="n">ci_upper</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">ci_lower_arr</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">arr</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Calculate NNH</span>
        <span class="n">nnt_or_nnh</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span> <span class="n">arr</span><span class="p">)</span>
        <span class="n">ci_lower</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span> <span class="n">ci_lower_arr</span><span class="p">)</span>  <span class="c1"># Invert CI for NNH</span>
        <span class="n">ci_upper</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span> <span class="n">ci_upper_arr</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Infinity&quot;</span><span class="p">,</span> <span class="s2">&quot;Infinity&quot;</span>  <span class="c1"># No effect, CI undefined</span>

    <span class="c1"># Handle undefined bounds when CI for ARR includes zero</span>
    <span class="k">if</span> <span class="n">ci_lower_arr</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ci_lower</span> <span class="o">=</span> <span class="s2">&quot;Infinity&quot;</span>
    <span class="k">if</span> <span class="n">ci_upper_arr</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ci_upper</span> <span class="o">=</span> <span class="s2">&quot;Infinity&quot;</span>

    <span class="k">return</span> <span class="n">nnt_or_nnh</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span>


<span class="c1"># Application with the Apixaban example</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span> <span class="c1"># Set 95% confidence interval</span>
<span class="n">nnt_or_nnh</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">calculate_nnt_ci</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">conf</span><span class="o">=</span><span class="n">conf</span><span class="p">)</span>

<span class="k">if</span> <span class="n">nnt_or_nnh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">metric_name</span> <span class="o">=</span> <span class="s2">&quot;NNT&quot;</span> <span class="k">if</span> <span class="n">nnt_or_nnh</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;NNH&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric_name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">nnt_or_nnh</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">conf</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">% confidence interval: (</span><span class="si">{</span><span class="n">ci_lower</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">ci_upper</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No treatment effect (ARR = 0)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NNT: 14
95% confidence interval: (11, 20)
</pre></div>
</div>
</div>
</div>
<p>When calculating the CI for NNT, it’s crucial to remember that if the CI for ARR includes zero, the CI for NNT will extend to infinity (or negative infinity if ARR is negative). This is because the NNT is undefined when there’s no effect (ARR = 0). In such cases, it’s common to report the upper or lower bound of the NNT CI, along with a statement indicating that the other bound is undefined.</p>
</section>
<section id="ci-for-rr">
<h4>CI for RR<a class="headerlink" href="#ci-for-rr" title="Link to this heading">#</a></h4>
<p>While the RR provides a valuable point estimate for the comparison of risks between two groups, it’s essential to acknowledge the inherent uncertainty associated with any estimate derived from sample data. To address this uncertainty, we turn to confidence intervals (CIs), which provide a range of plausible values for the true population RR.</p>
<p>Confidence intervals for RR are crucial for several reasons:</p>
<ul class="simple">
<li><p>Assessing precision: they indicate how precise our estimate of the RR is. A wider CI suggests greater uncertainty, while a narrower CI reflects a more precise estimate.</p></li>
<li><p>Statistical significance: if the CI for RR does not include 1 (the value indicating no association), it suggests a statistically significant association between the exposure and outcome.</p></li>
<li><p>Clinical relevance: CIs help gauge the practical importance of the observed association. A statistically significant RR might not be clinically meaningful if the CI includes values close to 1.</p></li>
</ul>
<p>The traditional formula for the CI of RR is <span class="math notranslate nohighlight">\(\text{CI} = \text{RR} \pm z^\ast \times s_\text{RR}\)</span>, where <span class="math notranslate nohighlight">\(\text{RR}\)</span> is the relative risk, <span class="math notranslate nohighlight">\(z^\ast\)</span> the critical z-value corresponding to the desired level of confidence (e.g., 1.96 for a 95% CI), and <span class="math notranslate nohighlight">\(s_\text{RR}\)</span> the standard error of the relative risk, often approximated as:</p>
<div class="math notranslate nohighlight">
\[s_\text{RR} \approx \text{RR} \times \sqrt{1/a \times b/n_1 + 1/c \times d/n_2}\]</div>
<p>However, a more accurate approach to calculating the CI for RR is based on the log transformation. This approach takes advantage of the fact that the logarithm of the RR tends to follow a normal distribution, especially for large sample sizes. Then we calculate the standard error of log(RR) as:</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{align}
s_{\log{\text{RR}}} &amp;= \sqrt{\mathrm{Var}(\log{\text{RR}})} \\
&amp;= \sqrt{\frac{1 - p_1}{a} + \frac{1 - p_2}{c}}
\end{align}
\end{split}\]</div>
<p><em>The formula for Var(logRR) is derived using a statistical technique called the <a class="reference external" href="https://en.wikipedia.org/wiki/Delta_method">Delta Method</a>. This method approximates the variance of a transformed variable (in this case, logRR) using the variance of the original variable (RR) and the derivative of the transformation function.</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_rr_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span><span class="p">):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;Calculates relative risk (RR) and its confidence interval (CI) from a 2x2 </span>
<span class="sd">contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table (np.ndarray): A 2x2 NumPy array representing the contingency table.</span>
<span class="sd">        conf (float, optional): Confidence level (default: 0.95).</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing:</span>
<span class="sd">            - RR (float): The relative risk.</span>
<span class="sd">            - ci_lower (float): Lower bound of the confidence interval.</span>
<span class="sd">            - ci_upper (float): Upper bound of the confidence interval.</span>
<span class="sd">    &quot;&quot;&quot;</span>
  <span class="k">if</span> <span class="n">table</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input table must be a 2x2 array.&quot;</span><span class="p">)</span>

  <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

  <span class="c1"># Risk calculations (AR = p2 - p1)</span>
  <span class="n">p1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span>
  <span class="n">p2</span> <span class="o">=</span> <span class="n">c</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>
  <span class="n">rr</span> <span class="o">=</span> <span class="n">p1</span> <span class="o">/</span> <span class="n">p2</span>

  <span class="c1"># Standard error of RR</span>
  <span class="n">log_rr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">rr</span><span class="p">)</span>
  <span class="n">se_log_rr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">((</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p1</span><span class="p">)</span> <span class="o">/</span> <span class="n">a</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p2</span><span class="p">)</span> <span class="o">/</span> <span class="n">c</span><span class="p">)</span>

  <span class="c1"># Z-score for desired confidence level</span>
  <span class="n">z</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">ppf</span><span class="p">((</span><span class="mi">1</span> <span class="o">+</span> <span class="n">conf</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

  <span class="c1"># Confidence interval on the log scale</span>
  <span class="n">ci_lower_log</span> <span class="o">=</span> <span class="n">log_rr</span> <span class="o">-</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_rr</span>
  <span class="n">ci_upper_log</span> <span class="o">=</span> <span class="n">log_rr</span> <span class="o">+</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_rr</span>
  <span class="n">ci_lower</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">ci_lower_log</span><span class="p">)</span>
  <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">ci_upper_log</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">rr</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span>


<span class="c1"># Application with the Apixaban example</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span> <span class="c1"># Set 95% confidence interval</span>
<span class="n">rr</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">calculate_rr_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="n">conf</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Relative risk (RR): </span><span class="si">{</span><span class="n">rr</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">conf</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">% confidence interval for RR: (</span><span class="si">{</span><span class="n">ci_lower</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">ci_upper</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Relative risk (RR): 5.283
95% confidence interval for RR: (3.007, 9.284)
</pre></div>
</div>
</div>
</div>
<p>Similarly to the CI for ARR, we can use the <a class="reference external" href="https://www.statsmodels.org/dev/generated/statsmodels.stats.proportion.confint_proportions_2indep.html"><code class="docutils literal notranslate"><span class="pre">statsmodels.stats.proportion.confint_proportions_2indep</span></code></a> function to calculate a reasonable approximation of other methods, such as the Koopman asymptotic score of the CI for the RR by setting <code class="docutils literal notranslate"><span class="pre">compare='ratio'</span></code> and <code class="docutils literal notranslate"><span class="pre">method='score'</span></code> arguments. This function also offers other methods, including ‘log’ (a variant of the Katz log method) and ‘log-adjusted’ (which aims to improve the ‘log’ method).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Application with the Apixaban example</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

<span class="c1"># Calculate the confidence interval using the score method</span>
<span class="n">ci_rr_approx</span> <span class="o">=</span> <span class="n">confint_proportions_2indep</span><span class="p">(</span>
    <span class="n">count1</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">nobs1</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">count2</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">nobs2</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span>
    <span class="n">compare</span><span class="o">=</span><span class="s1">&#39;ratio&#39;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;score&#39;</span><span class="p">,</span>  <span class="c1"># [&#39;log&#39;, &#39;log-adjusted&#39;, &#39;score&#39;]</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Approximate 95</span><span class="si">% c</span><span class="s2">onfidence interval for RR (&#39;score&#39; method):&quot;</span><span class="p">,</span>
    <span class="nb">tuple</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bound</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">bound</span> <span class="ow">in</span> <span class="n">ci_rr_approx</span><span class="p">)</span> <span class="c1"># type: ignore</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Approximate 95% confidence interval for RR (&#39;score&#39; method): (3.033, 9.231)
</pre></div>
</div>
</div>
</div>
<p>While these methods implemented in <code class="docutils literal notranslate"><span class="pre">confint_proportions_2indep</span></code> provide approximations to the Koopman and Katz methods, they are often sufficient for many practical purposes. For the most precise calculation using the Koopman asymptotic score or Katz log methods, we may need to explore alternative implementations or specialized packages, for example using the <a class="reference external" href="https://cran.r-project.org/web/packages/PropCIs/"><code class="docutils literal notranslate"><span class="pre">PropCIs</span></code> library in R</a> which computes two-sample confidence intervals for single, paired and independent proportions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">PropCIs</span><span class="p">)</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">n_exposed</span> <span class="o">&lt;-</span> <span class="mi">73</span><span class="o">+</span><span class="mi">756</span>
<span class="n">n_unexposed</span> <span class="o">&lt;-</span> <span class="mi">14</span><span class="o">+</span><span class="mi">826</span>
<span class="n">x_exposed</span> <span class="o">&lt;-</span> <span class="mi">73</span>
<span class="n">x_unexposed</span> <span class="o">&lt;-</span> <span class="mi">14</span>

<span class="c1"># Calculate the confidence interval using the score (Koopman) method</span>
<span class="n">ci</span> <span class="o">&lt;-</span> <span class="n">riskscoreci</span><span class="p">(</span><span class="n">x_exposed</span><span class="p">,</span> <span class="n">n_exposed</span><span class="p">,</span> <span class="n">x_unexposed</span><span class="p">,</span> <span class="n">n_unexposed</span><span class="p">,</span> <span class="n">conf</span><span class="o">.</span><span class="n">level</span><span class="o">=</span><span class="mf">0.95</span><span class="p">)</span>

<span class="c1"># Print the results</span>
<span class="n">cat</span><span class="p">(</span><span class="s2">&quot;95</span><span class="si">% c</span><span class="s2">onfidence interval for RR (Koopman): (&quot;</span><span class="p">,</span> <span class="n">ci</span><span class="p">[[</span><span class="s2">&quot;conf.int&quot;</span><span class="p">]][</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">ci</span><span class="p">[[</span><span class="s2">&quot;conf.int&quot;</span><span class="p">]][</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>95% confidence interval for RR (Koopman): ( 3.033904 , 9.229101 )
</pre></div>
</div>
</div>
</div>
</section>
<section id="ci-for-or">
<h4>CI for OR<a class="headerlink" href="#ci-for-or" title="Link to this heading">#</a></h4>
<p>Similar to relative risk (RR) and absolute risk reduction (ARR), the odds ratio (OR) also requires a confidence interval (CI) to accurately reflect the precision of our estimate. The CI for the OR provides a range of plausible values within which we can be reasonably confident that the true population OR lies.</p>
<p>Due to the non-normal distribution of the OR, especially when the odds are not close to 1, we typically use the log-transformed OR to calculate the CI. This is because log(OR) tends to be more normally distributed, making the CI calculation more reliable.</p>
<p>The formula for the confidence interval of the log odds ratio is <span class="math notranslate nohighlight">\(\text{CI}(\log\text{OR}) = \log\text{OR} \pm z^\ast \times s_{\log\text{OR}}\)</span>, where <span class="math notranslate nohighlight">\(\log\text{OR}\)</span> is the natural logarithm of the odds ratio, <span class="math notranslate nohighlight">\(z^\ast\)</span> is the critical z-value corresponding to the desired level of confidence (e.g., 1.96 for a 95% CI), and <span class="math notranslate nohighlight">\(s_{\log\text{OR}}\)</span> is the standard error of the log odds ratio, calculated as</p>
<div class="math notranslate nohighlight">
\[s_{\log \text{OR}} = \sqrt{1/a + 1/b + 1/c + 1/d}\]</div>
<p>where <span class="math notranslate nohighlight">\(a\)</span>, <span class="math notranslate nohighlight">\(b\)</span>, <span class="math notranslate nohighlight">\(c\)</span>, and <span class="math notranslate nohighlight">\(d\)</span> are the cell counts in the 2x2 contingency table.</p>
<p>Once the CI for <span class="math notranslate nohighlight">\(\log \text{OR}\)</span> is calculated, we can exponentiate the bounds to get the CI for the original OR, as:</p>
<div class="math notranslate nohighlight">
\[\text{CI(OR)} = [\exp{\text{lower bound} \log \text{OR}}, \exp{\text{upper bound} \log \text{OR}}]\]</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_or_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates odds ratio (OR) and its confidence interval (CI) from </span>
<span class="sd">    a 2x2 contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table (np.ndarray): A 2x2 NumPy array representing the contingency </span>
<span class="sd">        table.</span>
<span class="sd">        conf (float, optional): Confidence level (default: 0.95).</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing:</span>
<span class="sd">            - OR (float): The odds ratio.</span>
<span class="sd">            - ci_lower (float): Lower bound of the confidence interval.</span>
<span class="sd">            - ci_upper (float): Upper bound of the confidence interval.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">table</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input table must be a 2x2 array.&quot;</span><span class="p">)</span>

    <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

    <span class="c1"># Odds ratio calculation</span>
    <span class="n">odds_ratio</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span> <span class="o">*</span> <span class="n">d</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">b</span> <span class="o">*</span> <span class="n">c</span><span class="p">)</span>
    <span class="n">log_or</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">odds_ratio</span><span class="p">)</span>

    <span class="c1"># Standard error of log odds ratio</span>
    <span class="n">se_log_or</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">a</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="n">b</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="n">d</span><span class="p">)</span>

    <span class="c1"># Z-score for desired confidence level</span>
    <span class="n">z</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">ppf</span><span class="p">((</span><span class="mi">1</span> <span class="o">+</span> <span class="n">conf</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Confidence interval calculation</span>
    <span class="n">ci_lower</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">log_or</span> <span class="o">-</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_or</span><span class="p">)</span>
    <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">log_or</span> <span class="o">+</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_or</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">odds_ratio</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span> <span class="c1"># Set 95% confidence interval</span>

<span class="n">or_val</span><span class="p">,</span> <span class="n">ci_lower</span><span class="p">,</span> <span class="n">ci_upper</span> <span class="o">=</span> <span class="n">calculate_or_ci</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">conf</span><span class="o">=</span><span class="n">conf</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Odds ratio (OR): </span><span class="si">{</span><span class="n">or_val</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">conf</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">% confidence interval for OR: (</span><span class="si">{</span><span class="n">ci_lower</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">ci_upper</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Odds ratio (OR): 5.697
95% confidence interval for OR: (3.189, 10.178)
</pre></div>
</div>
</div>
</div>
<p>If the CI for the OR does not include 1, this suggests a statistically significant association between the exposure and outcome at the chosen confidence level.</p>
<p>Similar to the confidence interval for ARR and RR, we can utilize the versatile <code class="docutils literal notranslate"><span class="pre">statsmodels.stats.proportion.confint_proportions_2indep</span></code> function to obtain a reasonable approximation of the CI for the odds ratio (OR).</p>
<p>By setting the <code class="docutils literal notranslate"><span class="pre">compare='or'</span></code> and <code class="docutils literal notranslate"><span class="pre">method='score'</span></code> arguments, we can approximate the CI for the OR using a method similar to the Koopman asymptotic score method. The ‘score’ method here is an adaptation of the score test for the difference in proportions, providing a convenient approximation when the exact Koopman method might not be readily available.</p>
<p>Additionally, <code class="docutils literal notranslate"><span class="pre">confint_proportions_2indep</span></code> offers other methods for calculating the CI for OR:</p>
<ul class="simple">
<li><p>‘log’ method: a variant of the Katz log method that operates on the log scale but might not include small-sample corrections.</p></li>
<li><p>‘log-adjusted’ method: an improvement over the ‘log’ method that aims to provide better coverage probability, especially for smaller samples or extreme proportions.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Application with the Apixaban example</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

<span class="c1"># Calculate the confidence interval using the score method</span>
<span class="n">ci_rr_approx</span> <span class="o">=</span> <span class="n">confint_proportions_2indep</span><span class="p">(</span>
    <span class="n">count1</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">nobs1</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">count2</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">nobs2</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span>
    <span class="n">compare</span><span class="o">=</span><span class="s1">&#39;odds-ratio&#39;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;score&#39;</span><span class="p">,</span>  <span class="c1"># [&#39;logit&#39;, &#39;logit-adjusted&#39;, &#39;score&#39;]</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Approximate 95</span><span class="si">% c</span><span class="s2">onfidence interval for OR (&#39;score&#39; method):&quot;</span><span class="p">,</span>
    <span class="nb">tuple</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bound</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">bound</span> <span class="ow">in</span> <span class="n">ci_rr_approx</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Approximate 95% confidence interval for OR (&#39;score&#39; method): (3.212, 10.103)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="summary-table">
<h3>Summary table<a class="headerlink" href="#summary-table" title="Link to this heading">#</a></h3>
<section id="calculating-measures-of-association-and-effect-with-custom-functions">
<h4>Calculating measures of association and effect with custom functions<a class="headerlink" href="#calculating-measures-of-association-and-effect-with-custom-functions" title="Link to this heading">#</a></h4>
<p>Let’s create a comprehensive table summarizing the calculation of the P value, individual proportions, ARR, NNT, RR and OR, and their logarithms, and their respective CIs, incorporating the functions we’ve discussed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_metrics</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculates various metrics and confidence intervals from a </span>
<span class="sd">    2x2 contingency table.</span>

<span class="sd">    Args:</span>
<span class="sd">        table (np.ndarray): A 2x2 NumPy array representing the contingency table.</span>
<span class="sd">        conf (float, optional): Confidence level (default: 0.95).</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the calculated metrics and CIs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

    <span class="c1"># Proportions</span>
    <span class="n">p1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span>
    <span class="n">p2</span> <span class="o">=</span> <span class="n">c</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>

    <span class="c1"># Confidence intervals for proportions</span>
    <span class="n">ci_p1</span> <span class="o">=</span> <span class="n">proportion_confint</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">conf</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilson&#39;</span><span class="p">)</span>
    <span class="n">ci_p2</span> <span class="o">=</span> <span class="n">proportion_confint</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">conf</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilson&#39;</span><span class="p">)</span>

    <span class="c1"># Attributable risk and CI</span>
    <span class="n">arr</span><span class="p">,</span> <span class="n">ci_lower_arr</span><span class="p">,</span> <span class="n">ci_upper_arr</span> <span class="o">=</span> <span class="n">calculate_arr_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="p">)</span>

    <span class="c1"># Number needed to treat (NNT) or harm (NNH) and CI</span>
    <span class="n">nnt_or_nnh</span><span class="p">,</span> <span class="n">ci_lower_nnt</span><span class="p">,</span> <span class="n">ci_upper_nnt</span> <span class="o">=</span> <span class="n">calculate_nnt_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="p">)</span>

    <span class="c1"># Odds ratio and CI</span>
    <span class="n">odds_ratio</span><span class="p">,</span> <span class="n">ci_lower_or</span><span class="p">,</span> <span class="n">ci_upper_or</span> <span class="o">=</span> <span class="n">calculate_or_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="p">)</span>

    <span class="c1"># Relative risk and CI</span>
    <span class="n">rr</span><span class="p">,</span> <span class="n">ci_lower_rr</span><span class="p">,</span> <span class="n">ci_upper_rr</span> <span class="o">=</span> <span class="n">calculate_rr_ci</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">conf</span><span class="p">)</span>

    <span class="c1"># Log transformations and CIs</span>
    <span class="n">log_or</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">odds_ratio</span><span class="p">)</span>
    <span class="n">log_rr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">rr</span><span class="p">)</span>
    <span class="n">se_log_or</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">a</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="n">b</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="n">d</span><span class="p">)</span>
    <span class="n">se_log_rr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">((</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p1</span><span class="p">)</span><span class="o">/</span><span class="n">a</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p2</span><span class="p">)</span><span class="o">/</span><span class="n">c</span><span class="p">)</span>

    <span class="n">z</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">ppf</span><span class="p">((</span><span class="mi">1</span> <span class="o">+</span> <span class="n">conf</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>  

    <span class="n">ci_lower_log_or</span> <span class="o">=</span> <span class="n">log_or</span> <span class="o">-</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_or</span>
    <span class="n">ci_upper_log_or</span> <span class="o">=</span> <span class="n">log_or</span> <span class="o">+</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_or</span>

    <span class="n">ci_lower_log_rr</span> <span class="o">=</span> <span class="n">log_rr</span> <span class="o">-</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_rr</span>
    <span class="n">ci_upper_log_rr</span> <span class="o">=</span> <span class="n">log_rr</span> <span class="o">+</span> <span class="n">z</span> <span class="o">*</span> <span class="n">se_log_rr</span>
    
    <span class="c1"># Fisher&#39;s exact test</span>
    <span class="n">_</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>


    <span class="c1"># Create results DataFrame</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Metric&#39;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s1">&#39;P value&#39;</span><span class="p">,</span> <span class="s1">&#39;Proportion (control)&#39;</span><span class="p">,</span> <span class="s1">&#39;Proportion (treatment)&#39;</span><span class="p">,</span>
            <span class="s1">&#39;ARR&#39;</span><span class="p">,</span> <span class="s1">&#39;NNT/NNH&#39;</span><span class="p">,</span> <span class="s1">&#39;OR&#39;</span><span class="p">,</span> <span class="s1">&#39;log(OR)&#39;</span><span class="p">,</span> <span class="s1">&#39;RR&#39;</span><span class="p">,</span> <span class="s1">&#39;log(RR)&#39;</span><span class="p">],</span>
        <span class="s1">&#39;Estimate&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">p_value</span><span class="p">,</span> <span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">arr</span><span class="p">,</span> <span class="n">nnt_or_nnh</span><span class="p">,</span> <span class="n">odds_ratio</span><span class="p">,</span> <span class="n">log_or</span><span class="p">,</span> <span class="n">rr</span><span class="p">,</span> <span class="n">log_rr</span><span class="p">],</span>
        <span class="s1">&#39;CI lower&#39;</span><span class="p">:</span> <span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">ci_p1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ci_p2</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ci_lower_arr</span><span class="p">,</span> <span class="n">ci_lower_nnt</span><span class="p">,</span>
            <span class="n">ci_lower_or</span><span class="p">,</span> <span class="n">ci_lower_log_or</span><span class="p">,</span> <span class="n">ci_lower_rr</span><span class="p">,</span> <span class="n">ci_lower_log_rr</span><span class="p">],</span>
        <span class="s1">&#39;CI upper&#39;</span><span class="p">:</span> <span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">ci_p1</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">ci_p2</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">ci_upper_arr</span><span class="p">,</span> <span class="n">ci_upper_nnt</span><span class="p">,</span> 
            <span class="n">ci_upper_or</span><span class="p">,</span> <span class="n">ci_upper_log_or</span><span class="p">,</span> <span class="n">ci_upper_rr</span><span class="p">,</span> <span class="n">ci_upper_log_rr</span><span class="p">]</span>
    <span class="p">}</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># Round to 3 decimal places</span>

<span class="c1"># Application with the Apixaban example</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span> <span class="c1"># Set 95% confidence interval</span>

<span class="n">results_df</span> <span class="o">=</span> <span class="n">calculate_metrics</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">conf</span><span class="o">=</span><span class="n">conf</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results_df</span><span class="o">.</span><span class="n">to_markdown</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">numalign</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">stralign</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>| Metric                 | Estimate   | CI lower   | CI upper   |
|:-----------------------|:-----------|:-----------|:-----------|
| P value                | 0          | nan        | nan        |
| Proportion (control)   | 0.088      | 0.071      | 0.109      |
| Proportion (treatment) | 0.017      | 0.01       | 0.028      |
| ARR                    | 0.071      | 0.05       | 0.093      |
| NNT/NNH                | 14         | 11         | 20         |
| OR                     | 5.697      | 3.189      | 10.178     |
| log(OR)                | 1.74       | 1.16       | 2.32       |
| RR                     | 5.283      | 3.007      | 9.284      |
| log(RR)                | 1.665      | 1.101      | 2.228      |
</pre></div>
</div>
</div>
</div>
<p>This table provides a general overview. Specific functions or methods may vary depending on the chosen libraries and methods, i.e., approximations vs. exact methods, and the characteristics of the data.</p>
</section>
<section id="table2x2-for-risk-and-odds-ratio-analysis">
<h4>Table2x2 for risk and odds ratio analysis<a class="headerlink" href="#table2x2-for-risk-and-odds-ratio-analysis" title="Link to this heading">#</a></h4>
<p>In addition to the custom functions we’ve explored, the statsmodels library offers a convenient way to calculate various risk metrics and associated statistics, including confidence intervals, using the <a class="reference external" href="https://www.statsmodels.org/dev/generated/statsmodels.stats.contingency_tables.Table2x2.html"><code class="docutils literal notranslate"><span class="pre">Table2x2</span></code> class</a>. This class provides a comprehensive framework for analyzing 2x2 contingency tables, making it a valuable tool for comparing proportions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.contingency_tables</span> <span class="kn">import</span> <span class="n">Table2x2</span>

<span class="c1"># Assuming &#39;table&#39; is the 2x2 contingency table (NumPy array)</span>
<span class="n">table2x2</span> <span class="o">=</span> <span class="n">Table2x2</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>

<span class="c1"># Calculate and print the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">table2x2</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>               Estimate   SE   LCB   UCB   p-value
--------------------------------------------------
Odds ratio        5.697       3.189 10.178   0.000
Log odds ratio    1.740 0.296 1.160  2.320   0.000
Risk ratio        5.283       3.007  9.284   0.000
Log risk ratio    1.665 0.288 1.101  2.228   0.000
--------------------------------------------------
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="one-sided-test-application-to-toxicology-count-data">
<h3>One-sided test - Application to toxicology count data<a class="headerlink" href="#one-sided-test-application-to-toxicology-count-data" title="Link to this heading">#</a></h3>
<p>In toxicology studies, where the paramount goal is to assess the safety of a test article (TA), e.g., a new drug, minimizing the risk of false negatives - incorrectly concluding a TA is safe when it’s not - must be the central concern throughout the study’s design, analysis, and interpretation.</p>
<p>Data from toxicology studies frequently consist of event counts, such as the number of tumors observed in two treatment groups: treated with a TA and control. These counts can be easily organized into contingency tables, making Fisher’s exact test a relevant statistical tool for analyzing such data.</p>
<p>A common scenario is testing the hypothesis that the tumor rate in a treated group is equal to that in a control group, i.e., there is no treatment effect, against the alternative hypothesis that the tumor rate is higher in the treated group. We perform a <strong>one-sided test</strong> to specifically assess whether the odds of tumor development are significantly greater in the treated group compared to the control group (i.e., OR &gt; 1). Fisher’s exact test is particularly suitable when the number of observations in the contingency table is small, as it provides an exact P value without relying on large-sample approximations.</p>
<p>Although some statistical software packages may lack a dedicated function for Fisher’s exact test, it’s readily available in Python using the <code class="docutils literal notranslate"><span class="pre">fisher_exact</span></code> function from the SciPy package, as we’ve demonstrated earlier in this chapter.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">table_tumors</span><span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span>
        <span class="s1">&#39;treated&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">42</span><span class="p">],</span> <span class="c1"># exposed</span>
        <span class="s1">&#39;control&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">48</span><span class="p">]</span>  <span class="c1"># unexposed</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;tumor&#39;</span><span class="p">,</span> <span class="s1">&#39;no_tumor&#39;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">odds_ratio</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">table_tumors</span><span class="p">,</span> <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;greater&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Odds ratio</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">odds_ratio</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;P value (one-sided)&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p_value</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Odds ratio	 4.57
P value (one-sided) 0.046
</pre></div>
</div>
</div>
</div>
<p>A small P value (typically below 0.05) would provide evidence against the null hypothesis, suggesting that the tumor rate is indeed higher in the treated group.</p>
</section>
</section>
<section id="chi-squared-test">
<h2>Chi-squared test<a class="headerlink" href="#chi-squared-test" title="Link to this heading">#</a></h2>
<p>The chi-squared (<span class="math notranslate nohighlight">\(\chi^2\)</span>) test stands as a cornerstone in the analysis of <strong>categorical data</strong>. It’s a versatile statistical tool used to examine the relationship between two categorical variables, answering the question of whether there’s a significant association between them.</p>
<p>The chi-squared test’s versatility stems from its ability to handle various types of contingency tables, not just the 2x2 tables commonly analyzed with Fisher’s exact test. Whether we’re dealing with multiple groups, multiple outcomes, or a combination of both, the chi-squared test can adapt to the data structure.</p>
<p>This test is particularly powerful when:</p>
<ul class="simple">
<li><p>Comparing multiple groups: we want to assess whether the distribution of an outcome differs significantly across multiple groups or categories.</p></li>
<li><p>Investigating complex associations: we’re interested in exploring associations between multiple categorical variables, beyond simple 2x2 relationships.</p></li>
<li><p>Large sample sizes: when the sample size is sufficiently large, the chi-squared test provides a reliable way to assess the statistical significance of observed associations.</p></li>
</ul>
<section id="how-it-works">
<h3>How it works<a class="headerlink" href="#how-it-works" title="Link to this heading">#</a></h3>
<p>The core principle of the chi-squared test lies in comparing the <strong>observed</strong> frequencies in the contingency table to the <strong>expected</strong> frequencies if there were no association between the variables.</p>
<ul class="simple">
<li><p>Observed frequencies: the actual counts of individuals falling into each category combination in the data.</p></li>
<li><p>Expected frequencies: the counts we would expect to see in each category if the null hypothesis (no association) were true. These are calculated based on the marginal totals of the table.</p></li>
</ul>
<p>It quantifies the discrepancy between these two sets of frequencies and assesses whether this discrepancy is likely due to chance or reflects a genuine association. The key steps involved are:</p>
<ol class="arabic simple">
<li><p>Formulate hypotheses</p>
<ul class="simple">
<li><p>Null hypothesis (H0): there is no association between the two categorical variables.</p></li>
<li><p>Alternative hypothesis (H1): there is an association between the variables.</p></li>
</ul>
</li>
<li><p>Calculate expected frequencies: under the null hypothesis of independence, calculate the expected frequency for each cell in the contingency table.</p></li>
<li><p>Calculate chi-squared statistic: this statistic measures the overall discrepancy between observed and expected frequencies:</p></li>
</ol>
<div class="math notranslate nohighlight">
\[\chi^2 = \sum \frac{(\text{Observed} - \text{Expected})^2}{\text{Expected}}\]</div>
<ol class="arabic simple">
<li><p>Determine P value: the P value quantifies the probability of observing a chi-squared statistic as extreme as (or more extreme than) the one calculated, assuming the null hypothesis is true.</p></li>
<li><p>Interpret results: if the P value is small (typically below 0.05), we reject the null hypothesis and conclude that there is evidence of a statistically significant association between the variables.</p></li>
</ol>
</section>
<section id="example-of-the-mendel-s-peas">
<h3>Example of the Mendel’s peas<a class="headerlink" href="#example-of-the-mendel-s-peas" title="Link to this heading">#</a></h3>
<p>Gregor Mendel, the father of modern genetics, <a class="reference external" href="https://www.nature.com/scitable/topicpage/gregor-mendel-and-the-principles-of-inheritance-593/">meticulously studied the inheritance of traits in pea plants</a>. His experiments revealed fundamental principles of heredity that laid the groundwork for our understanding of genetics today. One of his key findings was that the traits of round/wrinkled seeds and yellow/green seeds followed predictable patterns of inheritance.</p>
<p>Mendel crossed purebred pea plants with round, yellow seeds (RRYY) with purebred plants with wrinkled, green seeds (rryy). The resulting first-generation (F1) plants all had round, yellow seeds (RrYy) due to the dominance of the round (R) and yellow (Y) alleles. When Mendel self-fertilized the F1 plants, he observed the following phenotypes in the second generation (F2):</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Phenotype</p></th>
<th class="head text-center"><p>Observed number</p></th>
<th class="head text-center"><p>Expected proportion</p></th>
<th class="head text-center"><p>Expected number</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>Round, Yellow</p></td>
<td class="text-center"><p>315</p></td>
<td class="text-center"><p>9/16</p></td>
<td class="text-center"><p>312.75</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Round, Green</p></td>
<td class="text-center"><p>108</p></td>
<td class="text-center"><p>3/16</p></td>
<td class="text-center"><p>104.25</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>Wrinkled, Yellow</p></td>
<td class="text-center"><p>101</p></td>
<td class="text-center"><p>3/16</p></td>
<td class="text-center"><p>104.25</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Wrinkled, Green</p></td>
<td class="text-center"><p>32</p></td>
<td class="text-center"><p>1/16</p></td>
<td class="text-center"><p>34.75</p></td>
</tr>
</tbody>
</table>
</div>
<p>The expected proportions in the table are derived directly from Mendel’s laws:</p>
<ul class="simple">
<li><p>Law of segregation: each parent (RrYy) can produce four types of gametes (RY, Ry, rY, ry) with equal probability due to the random segregation of alleles.</p></li>
<li><p>Law of independent assortment: the alleles for seed shape (R/r) and color (Y/y) are inherited independently. This results in 16 equally likely combinations of alleles in the offspring (RYRY, RYRy, etc.).</p></li>
<li><p>Dominance: the round (R) and yellow (Y) alleles are dominant. This means that only one copy of the dominant allele is needed to express the dominant trait.</p></li>
</ul>
<p>Combining these principles, we can deduce the expected proportions:</p>
<ul class="simple">
<li><p>Round, Yellow (RRYY, RrYY, RRYy, RrYy): 9 out of 16 possible combinations</p></li>
<li><p>Round, Green (RRyy, Rryy): 3 out of 16 combinations</p></li>
<li><p>Wrinkled, Yellow (rrYY, rrYy): 3 out of 16 combinations</p></li>
<li><p>Wrinkled, Green (rryy): 1 out of 16 combinations</p></li>
</ul>
</section>
<section id="chi-squared-goodness-of-fit-test">
<h3>Chi-squared goodness-of-fit test<a class="headerlink" href="#chi-squared-goodness-of-fit-test" title="Link to this heading">#</a></h3>
<p>To compare the experimental values (observed frequencies) in Mendel’s pea experiment to the expected proportions based on his laws of inheritance, we can use a <a class="reference external" href="https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.stats.chisquare.html"><code class="docutils literal notranslate"><span class="pre">scipy.stats.chisquare</span></code> chi-squared goodness-of-fit test</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chisquare</span>

<span class="c1"># Mendel&#39;s observed data</span>
<span class="n">observed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">315</span><span class="p">,</span> <span class="mi">108</span><span class="p">,</span> <span class="mi">101</span><span class="p">,</span> <span class="mi">32</span><span class="p">])</span>

<span class="c1"># Mendel&#39;s expected proportions</span>
<span class="n">expected_proportions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">9</span><span class="o">/</span><span class="mi">16</span><span class="p">,</span> <span class="mi">3</span><span class="o">/</span><span class="mi">16</span><span class="p">,</span> <span class="mi">3</span><span class="o">/</span><span class="mi">16</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="mi">16</span><span class="p">])</span>

<span class="c1"># Perform the chi-squared goodness-of-fit test</span>
<span class="n">chi2_stat</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">chisquare</span><span class="p">(</span>
    <span class="n">f_obs</span><span class="o">=</span><span class="n">observed</span><span class="p">,</span> <span class="n">f_exp</span><span class="o">=</span><span class="n">observed</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">*</span> <span class="n">expected_proportions</span><span class="p">)</span>

<span class="c1"># Print the results</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Observed frequencies:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">observed</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Expected frequencies:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">((</span><span class="n">observed</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">*</span> <span class="n">expected_proportions</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chi-squared statistic:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">chi2_stat</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>

<span class="c1"># Interpretation</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="k">if</span> <span class="n">p_value</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reject the null hypothesis (P &lt; </span><span class="si">{</span><span class="n">alpha</span><span class="si">}</span><span class="s2">). </span><span class="se">\</span>
<span class="s2">The observed frequencies do not fit the expected proportions.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">fr</span><span class="s2">&quot;Fail to reject the null hypothesis (P = </span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">). </span><span class="se">\</span>
<span class="s2">The observed frequencies are consistent with the expected proportions.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed frequencies:
[315 108 101  32]
Expected frequencies:
[312.75 104.25 104.25  34.75]
Chi-squared statistic: 0.47
Fail to reject the null hypothesis (P = 0.9254). \
The observed frequencies are consistent with the expected proportions.
</pre></div>
</div>
</div>
</div>
<p>With such a large P value, the deviations between the observed frequencies and the expected frequencies based on Mendel’s theory are small and consistent with random variation. This indicates that Mendel’s observed data support his genetic theory of inheritance.</p>
<section id="the-chi-squared-distribution">
<h4>The chi-squared distribution<a class="headerlink" href="#the-chi-squared-distribution" title="Link to this heading">#</a></h4>
<p>The chi-squared statistic calculated from a contingency table follows a chi-squared distribution, which can be used to determine the P value associated with this statistic. Like the t-distribution, the shape of the chi-squared distribution depends on the degrees of freedom.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chi2</span>

<span class="c1"># Degrees of freedom to plot</span>
<span class="n">degrees_of_freedom</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">]</span>

<span class="c1"># Generate x values for the distribution</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>  <span class="c1"># Range from 0 to 20 to capture most of the distribution</span>

<span class="c1"># Plot distributions for each degree of freedom</span>
<span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">degrees_of_freedom</span><span class="p">:</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">chi2</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">df</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;DF=</span><span class="si">{</span><span class="n">df</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># Add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;χ²&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Probability density&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;χ² distribution for various degrees of freedom&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Customize plot aesthetics</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">.1</span><span class="p">,</span> <span class="mf">.2</span><span class="p">,</span> <span class="mf">.3</span><span class="p">,</span> <span class="mf">.4</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c65a6f0c0c81e33ffc4255c70d14ad9e9b15dc8de30926c6731f234168628b15.png" src="_images/c65a6f0c0c81e33ffc4255c70d14ad9e9b15dc8de30926c6731f234168628b15.png" />
</div>
</div>
<p>The plot illustrates that the chi-squared distribution is right-skewed (long tail to the right). Moreover, as the degrees of freedom increase, the distribution becomes more symmetrical and approaches a normal distribution.</p>
</section>
<section id="degrees-of-freedom-in-goodness-of-fit-tests">
<h4>Degrees of freedom in goodness-of-fit tests<a class="headerlink" href="#degrees-of-freedom-in-goodness-of-fit-tests" title="Link to this heading">#</a></h4>
<p>In a chi-squared goodness-of-fit test, the degrees of freedom are simply the number of categories minus 1. In the case of Mendel’s pea experiment, there are 4 categories (round yellow, round green, wrinkled yellow, wrinkled green), so the degrees of freedom would be 3. They represent the number of categories whose frequencies can vary freely, given the constraint that the total number of observations is fixed. Once we know the frequencies of all but one category, the frequency of the remaining category is automatically determined.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Degrees of freedom for Mendel&#39;s example</span>
<span class="n">Df</span> <span class="o">=</span> <span class="mi">3</span>

<span class="c1"># Significance level (alpha)</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span>

<span class="c1"># Calculate critical value</span>
<span class="n">chi2_crit</span> <span class="o">=</span> <span class="n">chi2</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">Df</span><span class="p">)</span>

<span class="c1"># Calculate P value using chi-squared stat from previous calculation</span>
<span class="n">p_value</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">chi2</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">chi2_stat</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">Df</span><span class="p">)</span>

<span class="c1"># Generate x values for plotting</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">hx</span> <span class="o">=</span> <span class="n">chi2</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">Df</span><span class="p">)</span> 

<span class="c1"># Create the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">hx</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>

<span class="c1"># Plot the critical value</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">chi2_crit</span><span class="p">,</span> <span class="c1"># type: ignore</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orangered&#39;</span><span class="p">,</span>
    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;χ²* (</span><span class="si">{</span><span class="n">chi2_crit</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Shade the probability alpha</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span>
    <span class="n">x</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_crit</span><span class="p">],</span> <span class="n">hx</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_crit</span><span class="p">],</span>
    <span class="n">linestyle</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;tomato&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;α (</span><span class="si">{</span><span class="n">alpha</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Plot the observed chi-squared statistic</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">chi2_stat</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;limegreen&#39;</span><span class="p">,</span>
    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;χ² (</span><span class="si">{</span><span class="n">chi2_stat</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Shade the P value area</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span>
    <span class="n">x</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_stat</span><span class="p">],</span> <span class="n">hx</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_stat</span><span class="p">],</span>
    <span class="c1"># hatch=&#39;///&#39;, edgecolor=&quot;limegreen&quot;, facecolor=&#39;lime&#39;, alpha=.5,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;greenyellow&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P (</span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;χ²&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">margins</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">&#39;χ² distribution (DF=</span><span class="si">{</span><span class="n">Df</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">();</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/09b43271846222dd39195895ee6ed00133dbee8cc348d06600afa687d6e0e865.png" src="_images/09b43271846222dd39195895ee6ed00133dbee8cc348d06600afa687d6e0e865.png" />
</div>
</div>
<p>While the chi-squared goodness-of-fit test is inherently <strong>one-tailed</strong> due to the nature of the chi-squared distribution and test statistic, the interpretation can be either directional or non-directional depending on the specific hypothesis.</p>
</section>
</section>
<section id="chi-squared-test-of-independence">
<h3>Chi-squared test of independence<a class="headerlink" href="#chi-squared-test-of-independence" title="Link to this heading">#</a></h3>
<p>In this specific case, both seed shape and seed color have two categories each, resulting in a 2x2 (two rows by two columns) contingency table. The chi-squared test is often used to analyze 2x2 tables to determine if there is a significant association between the two categorical variables. Chi-squared tests can be applied to tables with more than two rows or columns, allowing us to analyze relationships between variables with multiple categories. In some cases, we might have a single row representing a categorical variable with multiple categories. We can still use the chi-squared test to compare the observed frequencies in each category to expected frequencies based on some theoretical distribution.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p>Round seeds</p></th>
<th class="head"><p>Wrinkled seeds</p></th>
<th class="head"><p>Total</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yellow seeds</p></td>
<td><p>315</p></td>
<td><p>101</p></td>
<td><p>416</p></td>
</tr>
<tr class="row-odd"><td><p>Green seeds</p></td>
<td><p>108</p></td>
<td><p>32</p></td>
<td><p>140</p></td>
</tr>
<tr class="row-even"><td><p>Total</p></td>
<td><p>423</p></td>
<td><p>133</p></td>
<td><p>556</p></td>
</tr>
</tbody>
</table>
</div>
<section id="mosaic-plot">
<h4>Mosaic plot<a class="headerlink" href="#mosaic-plot" title="Link to this heading">#</a></h4>
<p>The <a class="reference external" href="https://www.statsmodels.org/stable/generated/statsmodels.graphics.mosaicplot.mosaic.html"><code class="docutils literal notranslate"><span class="pre">mosaic</span></code> function within the <code class="docutils literal notranslate"><span class="pre">statsmodels.graphics.mosaicplot</span></code> module</a> creates a mosaic plot, a powerful graphical representation of contingency tables that visualizes the relative frequencies of combinations of categories within a contingency table, the area of each rectangle in the plot being directly <em>proportional to the frequency</em> of the corresponding category combination. Mosaic plots can effectively display relationships between multiple categorical variables, revealing hierarchical relationships between variables, and highlighting how the distribution of one variable changes depending on the levels of another variable.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.graphics.mosaicplot</span> <span class="kn">import</span> <span class="n">mosaic</span>

<span class="c1"># Mendel&#39;s observed data (2x2 contingency table)</span>
<span class="n">observed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="p">[</span><span class="mi">315</span><span class="p">,</span> <span class="mi">101</span><span class="p">],</span>  <span class="c1"># Yellow, round and wrinkled</span>
        <span class="p">[</span><span class="mi">108</span><span class="p">,</span> <span class="mi">32</span><span class="p">]</span>    <span class="c1"># Green, round and wrinkled</span>
    <span class="p">])</span>

<span class="c1"># Create a DataFrame from the observed data</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">observed</span><span class="p">,</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Round&#39;</span><span class="p">,</span> <span class="s1">&#39;Wrinkled&#39;</span><span class="p">],</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Yellow&#39;</span><span class="p">,</span> <span class="s1">&#39;Green&#39;</span><span class="p">])</span>

<span class="c1"># Create a mosaic plot</span>
<span class="n">mosaic</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">stack</span><span class="p">())</span>  <span class="c1"># Stacking the DataFrame for proper mosaic plot input</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Mosaic plot of Mendel&#39;s pea experiment&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/1d27c457d50e484c4ce565e5b037a0f061fb493ff872ab6aaa47e50abc4f9255.png" src="_images/1d27c457d50e484c4ce565e5b037a0f061fb493ff872ab6aaa47e50abc4f9255.png" />
</div>
</div>
</section>
<section id="interpreting-the-results-of-the-chi-squared-test-of-independence">
<h4>Interpreting the results of the chi-squared test of independence<a class="headerlink" href="#interpreting-the-results-of-the-chi-squared-test-of-independence" title="Link to this heading">#</a></h4>
<p>The <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html"><code class="docutils literal notranslate"><span class="pre">scipy.stats.chi2_contingency</span></code> function</a> performs the chi-squared test of independence for contingency tables. It’s designed to assess whether there is a significant association between two categorical variables. It takes a contingency table (a 2D array or list of lists) as input, representing the observed frequencies of the categories. It returns the χ² test statistic, the P value of the test, the degrees of freedom and the expected frequencies under the null hypothesis of independence.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">chi2_contingency</span></code> function doesn’t inherently know about Mendel’s laws of inheritance or any specific biological mechanism. It calculates expected values based on a purely statistical assumption: independence between the categorical variables. The function first calculates the row and column totals of the observed contingency table. These totals represent the <strong>marginal frequencies</strong> of each category for each variable. It then calculates the expected proportion for each cell in the table under the assumption of independence. This is done by multiplying the corresponding row proportion and column proportion and dividing by the total number of observations. Finally, the expected frequencies are obtained by multiplying the expected proportions by the total number of observations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chi2_contingency</span>

<span class="c1"># Perform the chi-squared test of independence on the previous contingency table</span>
<span class="n">chi2_stat</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">dof</span><span class="p">,</span> <span class="n">expected</span> <span class="o">=</span> <span class="n">chi2_contingency</span><span class="p">(</span>
    <span class="n">observed</span><span class="o">=</span><span class="n">observed</span><span class="p">,</span>
    <span class="n">lambda_</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">,</span>  <span class="c1"># could also use &#39;cressie-read&#39;, &#39;log-likelihood&#39;, &#39;neyman&#39;, etc.</span>
    <span class="n">correction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>    <span class="c1"># Yates&#39; correction for continuity</span>
<span class="p">)</span>

<span class="c1"># Print the results</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Observed Frequencies:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">observed</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Expected Frequencies (under null hypothesis of independence):&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expected</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Chi-Squared Statistic:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">chi2_stat</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Degrees of Freedom:&quot;</span><span class="p">,</span> <span class="n">dof</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;P value:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p_value</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span> <span class="c1"># type: ignore</span>

<span class="c1"># Interpret the results (using alpha = 0.05 as significance level)</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="k">if</span> <span class="n">p_value</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">:</span> <span class="c1"># type: ignore</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Reject the null hypothesis. There is a significant </span><span class="se">\</span>
<span class="s2">association between seed shape and color.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Fail to reject the null hypothesis. </span><span class="se">\n</span><span class="s2">There is not enough </span><span class="se">\</span>
<span class="s2">evidence to suggest an association between seed shape and color.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed Frequencies:
[[315 101]
 [108  32]]

Expected Frequencies (under null hypothesis of independence):
[[316.49  99.51]
 [106.51  33.49]]

Chi-Squared Statistic: 0.0513
Degrees of Freedom: 1
P value: 0.8208

Fail to reject the null hypothesis. 
There is not enough evidence to suggest an association between seed shape and color.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Degrees of freedom for Mendel&#39;s example</span>
<span class="n">Df</span> <span class="o">=</span> <span class="n">dof</span>

<span class="c1"># Significance level (alpha)</span>
<span class="n">α</span> <span class="o">=</span> <span class="mf">0.05</span>

<span class="c1"># Calculate critical value</span>
<span class="n">chi2_crit</span> <span class="o">=</span> <span class="n">chi2</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">α</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">Df</span><span class="p">)</span>

<span class="c1"># Calculate P value using chi-squared stat from previous calculation</span>
<span class="n">p_value</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">chi2</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">chi2_stat</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">Df</span><span class="p">)</span>

<span class="c1"># Generate x values for plotting</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">hx</span> <span class="o">=</span> <span class="n">chi2</span><span class="o">.</span><span class="n">pdf</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="n">Df</span><span class="p">)</span> 

<span class="c1"># Create the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">hx</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>

<span class="c1"># Plot the critical value</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">chi2_crit</span><span class="p">,</span> <span class="c1"># type: ignore</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orangered&#39;</span><span class="p">,</span>
    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;χ²* (</span><span class="si">{</span><span class="n">chi2_crit</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Shade the probability alpha</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span>
    <span class="n">x</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_crit</span><span class="p">],</span> <span class="n">hx</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_crit</span><span class="p">],</span>
    <span class="n">linestyle</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;tomato&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;α (</span><span class="si">{</span><span class="n">α</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Plot the observed chi-squared statistic</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">chi2_stat</span><span class="p">,</span> <span class="c1"># type: ignore</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;limegreen&#39;</span><span class="p">,</span>
    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;χ² (</span><span class="si">{</span><span class="n">chi2_stat</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Shade the P value area</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span>
    <span class="n">x</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_stat</span><span class="p">],</span> <span class="n">hx</span><span class="p">[</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="n">chi2_stat</span><span class="p">],</span> <span class="c1"># type: ignore</span>
    <span class="c1"># hatch=&#39;///&#39;, edgecolor=&quot;limegreen&quot;, facecolor=&#39;lime&#39;, alpha=.5,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;greenyellow&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P (</span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

<span class="c1"># Add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;χ²&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">margins</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mf">.5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span>
    <span class="sa">f</span><span class="s1">&#39;χ² distribution (DF=</span><span class="si">{</span><span class="n">Df</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">();</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/298fa9707bea34e1a188a45ef7c0c3701ea876a56b667d1db0e91c65f6fbab2c.png" src="_images/298fa9707bea34e1a188a45ef7c0c3701ea876a56b667d1db0e91c65f6fbab2c.png" />
</div>
</div>
<p>While the chi-squared test is widely used, it’s important to be aware of its limitations and assumptions:</p>
<ul class="simple">
<li><p>Large sample size: the chi-squared test relies on <em>large-sample approximations</em> to the <strong>chi-squared distribution</strong>. It’s generally considered reliable when the expected frequencies in each cell of the contingency table are <em>at least 5</em>. For smaller sample sizes or sparse tables, Fisher’s exact test might be a more appropriate choice.</p></li>
<li><p>Independence: the observations in the contingency table must be <strong>independent</strong>. This means that each individual should contribute only one count to the table, and the occurrence of one event should not influence the occurrence of another event.</p></li>
</ul>
<p><u>Note:</u> when there are only two categories, the binomial test can compute exact P values without relying on large-sample approximations. This makes it particularly useful for small to moderate sample sizes. However, for very large sample sizes, computational considerations might make the normal approximation to the binomial distribution a more practical choice, as long as the conditions for this approximation are met.</p>
</section>
<section id="degrees-of-freedom-in-independence-tests">
<h4>Degrees of freedom in independence tests<a class="headerlink" href="#degrees-of-freedom-in-independence-tests" title="Link to this heading">#</a></h4>
<p>In a chi-squared test of independence, the degrees of freedom represent the number of cells in the contingency table that can vary freely <em>once the marginal totals (row and column sums) are fixed</em>. It’s a measure of how much flexibility or “wiggle room” the data has within the constraints of the fixed totals. For a contingency table with r rows and c columns, the degrees of freedom are calculated as DF = (r - 1) * (c - 1).</p>
<p>In Mendel’s pea experiment, while there are 4 phenotype categories, the underlying genetic model (his laws of inheritance) imposes constraints on how the data can be distributed. Once we know the total number of plants and the expected proportions based on Mendel’s laws, we can determine the expected frequency for three of the cells. The fourth cell’s expected frequency is then automatically determined to maintain the correct marginal totals.</p>
<p>We have a 2x2 table (2 rows for seed color and 2 columns for seed shape), so the degrees of freedom are 1. Imagine we have a 2x2 table with the marginal totals, once we fill in one cell, the rest of the cells are automatically determined due to the fixed marginal totals. We only <em>had the freedom</em> to choose the value of one cell; the rest were constrained. Hence, the degrees of freedom in this case are 1.</p>
<p>The degrees of freedom in a chi-squared test reflect the number of independent pieces of information in the data that are available to estimate the parameters of interest.  It’s not simply the number of categories, but rather the number of categories that can vary freely given the constraints imposed by the marginal totals and any underlying model.</p>
</section>
<section id="yates-correction">
<h4>Yates’ correction<a class="headerlink" href="#yates-correction" title="Link to this heading">#</a></h4>
<p><a class="reference external" href="https://scirp.org/reference/referencespapers?referenceid=3172298">Yates’ correction</a> adjusts the chi-squared test statistic by <em>subtracting 0.5 from the absolute difference between observed and expected frequencies</em> in each cell of the contingency table, making it a more conservative test and improving its accuracy for small sample sizes.</p>
<p>The chi-squared test relies on the assumption that the sampling distribution of the test statistic approximates a chi-squared distribution. However, this approximation can be inaccurate when dealing with small sample sizes or sparse contingency tables (where some expected frequencies are low). Yates’ correction helps mitigate this issue by slightly reducing the chi-squared statistic, making it more closely match the theoretical chi-squared distribution under these conditions.</p>
</section>
<section id="directional-hypothesis">
<h4>Directional hypothesis<a class="headerlink" href="#directional-hypothesis" title="Link to this heading">#</a></h4>
<p>The chi-squared test of independence is <strong>not symmetrical</strong> with respect to the rows and columns of the table, it typically has an <em>implicit directional hypothesis</em>. In Mendel’s pea experiment, we were implicitly testing whether seed shape influences seed color (or vice versa). When we transpose the table, we’re essentially reversing this directionality.</p>
<p>The expected frequencies calculated under the null hypothesis of independence change when we transpose the table. These expected frequencies are based on the marginal totals (row and column sums), which change when we swap rows and columns. The interpretation of the results can also be affected. If we transpose the table, we might be testing a different hypothesis than the original one.</p>
<p>Before performing the chi-squared test, clearly define the research question and hypothesis. This will help decide which variable should be represented by rows and which by columns in the contingency table. In some cases, it might be insightful to analyze both the original and transposed table to explore potential relationships in both directions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform the chi-squared test of independence on the table transposed</span>
<span class="n">chi2_stat</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">dof</span><span class="p">,</span> <span class="n">expected</span> <span class="o">=</span> <span class="n">chi2_contingency</span><span class="p">(</span><span class="n">observed</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

<span class="c1"># Print the results</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Observed Frequencies:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">observed</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Expected Frequencies (under null hypothesis of independence):&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expected</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Chi-Squared Statistic:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">chi2_stat</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span> <span class="c1"># type: ignore</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Degrees of Freedom:&quot;</span><span class="p">,</span> <span class="n">dof</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;P value:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p_value</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed Frequencies:
[[315 108]
 [101  32]]

Expected Frequencies (under null hypothesis of independence):
[[316.49 106.51]
 [ 99.51  33.49]]

Chi-Squared Statistic: 0.0513
Degrees of Freedom: 1
P value: 0.8208
</pre></div>
</div>
</div>
</div>
</section>
<section id="chi-squared-independence-tests-using-pingouin">
<h4>Chi-squared independence tests using Pingouin<a class="headerlink" href="#chi-squared-independence-tests-using-pingouin" title="Link to this heading">#</a></h4>
<p>Let’s illustrate a concrete example using <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html"><code class="docutils literal notranslate"><span class="pre">pd.crosstab</span></code> to create a cross tabulation / contingency table</a> from a <a class="reference external" href="https://link.springer.com/article/10.1007/s12046-018-0916-9">case study dataset on the heart diseases</a> provided by Pingouin, and then analyze if gender is a good categorical predictor for the presence of heart disease using the <a class="reference external" href="https://pingouin-stats.org/build/html/generated/pingouin.chi2_independence.html"><code class="docutils literal notranslate"><span class="pre">chi2_independence</span></code> function from the Pingouin library</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pingouin</span> <span class="k">as</span> <span class="nn">pg</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pg</span><span class="o">.</span><span class="n">read_dataset</span><span class="p">(</span><span class="s1">&#39;chi2_independence&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;DataFrame&#39;s head:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">---------------------------------&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Contingency table:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sex&#39;</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;target&#39;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DataFrame&#39;s head:
    age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  

---------------------------------
Contingency table:
 target    0   1
sex            
0        24  72
1       114  93
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform chi-squared test using pingouin</span>
<span class="n">expected</span><span class="p">,</span> <span class="n">observed</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="n">pg</span><span class="o">.</span><span class="n">chi2_independence</span><span class="p">(</span>
    <span class="n">data</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;sex&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;target&#39;</span><span class="p">,</span>
    <span class="n">correction</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># No Yates&#39; correction for illustration</span>
<span class="p">)</span>

<span class="c1"># Print the results</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Observed Frequencies:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">observed</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Expected Frequencies (under null hypothesis of independence):&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expected</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Chi-Squared Statistic:&quot;</span><span class="p">)</span>
<span class="n">pg</span><span class="o">.</span><span class="n">print_table</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed Frequencies:
target    0   1
sex            
0        24  72
1       114  93

Expected Frequencies (under null hypothesis of independence):
target      0       1
sex                  
0       43.72   52.28
1       94.28  112.72

Chi-Squared Statistic:
test                  lambda    chi2    dof    pval    cramer    power
------------------  --------  ------  -----  ------  --------  -------
pearson                1.000  23.914  1.000   0.000     0.281    0.998
cressie-read           0.667  24.149  1.000   0.000     0.282    0.998
log-likelihood         0.000  24.841  1.000   0.000     0.286    0.999
freeman-tukey         -0.500  25.576  1.000   0.000     0.290    0.999
mod-log-likelihood    -1.000  26.526  1.000   0.000     0.296    0.999
neyman                -2.000  29.205  1.000   0.000     0.310    1.000
</pre></div>
</div>
</div>
</div>
<p>The output from <code class="docutils literal notranslate"><span class="pre">pingouin.chi2_independence</span></code> includes the chi-squared statistic, degrees of freedom, and P value for each test. A small P value (typically &lt; 0.05) indicates evidence against the null hypothesis of independence, suggesting a statistically significant association between the variables. A larger chi-squared value generally indicates a stronger association, but its magnitude should be interpreted in conjunction with the degrees of freedom.</p>
<p><code class="docutils literal notranslate"><span class="pre">pingouin.chi2_independence</span></code> doesn’t just perform the standard Pearson’s chi-squared test. It actually computes several related chi-squared tests, each with slightly different theoretical underpinnings and assumptions. This allows us to compare the results of different tests and choose the one most appropriate for the specific data and analysis goals.</p>
<ol class="arabic simple">
<li><p>Pearson’s chi-squared test (<code class="docutils literal notranslate"><span class="pre">pearson</span></code>): this is the classic chi-squared test of independence that we’re likely most familiar with. It compares the observed frequencies in the contingency table to the expected frequencies under the null hypothesis of independence. It’s the most commonly used chi-squared test and is generally suitable for most situations.</p></li>
<li><p>Cressie-Read power divergence statistic (<code class="docutils literal notranslate"><span class="pre">cressie-read</span></code>): this is a family of chi-squared statistics that includes Pearson’s chi-squared as a special case (when the lambda parameter is 1). By varying the lambda parameter, we can adjust the test’s sensitivity to different types of deviations from independence. Pingouin calculates this statistic for several lambda values (2/3, 0, -1/2, -1, -2), allowing us to compare the results.</p></li>
<li><p>Log-likelihood ratio test (<code class="docutils literal notranslate"><span class="pre">log-likelihood</span></code>): this test compares the likelihood of the observed data under the null hypothesis of independence to the likelihood under a model where the variables are allowed to be associated. It’s an alternative to the Pearson chi-squared test and is often preferred for smaller sample sizes.</p></li>
<li><p>Freeman-Tukey statistic (<code class="docutils literal notranslate"><span class="pre">freeman-tukey</span></code>): this test is similar to the Pearson chi-squared, but it uses a variance-stabilizing transformation to make the test statistic more closely follow a chi-squared distribution, particularly for small sample sizes.</p></li>
<li><p>Neyman modified chi-squared (<code class="docutils literal notranslate"><span class="pre">neyman</span></code>): this test is another modification of the Pearson chi-squared test that aims to improve its performance for small sample sizes or sparse tables.</p></li>
</ol>
</section>
</section>
</section>
<section id="cochran-armitage-test-for-trend">
<h2>Cochran-Armitage test for trend<a class="headerlink" href="#cochran-armitage-test-for-trend" title="Link to this heading">#</a></h2>
<p>While the chi-squared test assesses general associations between <em>categorical variables</em>, the <a class="reference external" href="https://en.wikipedia.org/wiki/Cochran%E2%80%93Armitage_test_for_trend">Cochran-Armitage trend test (also known as the trend test)</a> is a specialized tool designed to detect linear trends in proportions across <strong>ordinal categories</strong>. This test is particularly valuable in biostatistics, especially in toxicology and dose-response studies, where we often want to determine if there’s a relationship between increasing doses or exposure levels and the probability of a specific outcome.</p>
<section id="how-the-trend-test-works">
<h3>How the trend test works<a class="headerlink" href="#how-the-trend-test-works" title="Link to this heading">#</a></h3>
<p>The trend test essentially assigns scores to each category of the ordinal variable (e.g., 1, 2, 3 for low, medium, high dose) and then calculates a weighted chi-squared statistic. This weighted statistic takes into account the ordering of the categories, making it sensitive to linear trends.</p>
<p>The trend test is most appropriate when:</p>
<ul class="simple">
<li><p>Ordinal categories: the data involves a binary outcome (e.g., presence/absence of disease) and an explanatory variable with ordered categories (e.g., increasing dose levels: low, medium, high).</p></li>
<li><p>Linear trend suspected: we suspect that the proportion of the outcome changes in a linear fashion across the ordered categories. For example, we might hypothesize that the risk of disease increases with higher doses of a toxin.</p></li>
</ul>
<p>The trend test is more powerful than the standard chi-squared test when a linear trend truly exists. This means it’s more likely to detect a real association when one is present. Moreover, it specifically tests for a linear trend, making it more focused than the general chi-squared test, which only assesses whether there’s any association. Finally, the trend test is ideally suited for dose-response studies, where the goal is to understand how the effect of a substance changes with increasing exposure levels.</p>
<p>Note that the Cochran-Armitage test assumes that the trend is consistent across dose groups (i.e., the effect of each dose increment is roughly the same).</p>
</section>
<section id="example-of-varying-the-dose-of-a-drug">
<h3>Example of varying the dose of a drug<a class="headerlink" href="#example-of-varying-the-dose-of-a-drug" title="Link to this heading">#</a></h3>
<p>We will use the (two-sided) Cochran-Armitage trend test in Python with <a class="reference external" href="https://www.statsmodels.org/dev/generated/statsmodels.stats.contingency_tables.Table.test_ordinal_association.html"><code class="docutils literal notranslate"><span class="pre">statsmodels.stats.contingency_tables.Table.test_ordinal_association</span></code></a>. Let’s assume we have the following data from a toxicology study, where we exposed different groups of animals to increasing doses of a test article (TA) and recorded the number of tumors:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p>0 dose (control)</p></th>
<th class="head"><p>1 dose (low)</p></th>
<th class="head"><p>2 doses (medium)</p></th>
<th class="head"><p>3 doses (high)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Animals with tumor</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>3</p></td>
<td><p>6</p></td>
</tr>
<tr class="row-odd"><td><p>Total animals examined</p></td>
<td><p>50</p></td>
<td><p>50</p></td>
<td><p>50</p></td>
<td><p>50</p></td>
</tr>
</tbody>
</table>
</div>
<p>Note that in this example, the total number of animals per dose group is the same. This does not have to be the case for the Cochran-Armitage test.</p>
<section id="table-vs-table2x2">
<h4><code class="docutils literal notranslate"><span class="pre">Table</span></code> vs. <code class="docutils literal notranslate"><span class="pre">Table2x2</span></code><a class="headerlink" href="#table-vs-table2x2" title="Link to this heading">#</a></h4>
<p>Note the <a class="reference external" href="https://www.statsmodels.org/stable/contingency_tables.html">distinction between <code class="docutils literal notranslate"><span class="pre">statsmodels.Table</span></code> and <code class="docutils literal notranslate"><span class="pre">statsmodels.Table2x2</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.statsmodels.org/stable/generated/statsmodels.stats.contingency_tables.Table.html#statsmodels.stats.contingency_tables.Table"><code class="docutils literal notranslate"><span class="pre">statsmodels.Table</span></code></a>:</p>
<ul>
<li><p>General purpose: this class is designed for working with contingency tables of <em>any size</em> (i.e., tables with more than two rows and/or columns).</p></li>
<li><p>Functionality: it offers a wide range of methods for analyzing contingency tables, including:</p>
<ul>
<li><p>Chi-squared test of independence</p></li>
<li><p>Fisher’s exact test</p></li>
<li><p>Measures of association (e.g., Cramer’s V, contingency coefficient)</p></li>
<li><p>Tests for marginal homogeneity</p></li>
<li><p>Calculation of expected frequencies, residuals, etc.</p></li>
</ul>
</li>
<li><p>Flexibility: it can handle various types of input data, including raw data arrays, frequency counts, and proportions.</p></li>
<li><p>When to use: choose <code class="docutils literal notranslate"><span class="pre">Table</span></code> when we have a contingency table that is not strictly 2x2 and need to perform a variety of analyses.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://www.statsmodels.org/stable/generated/statsmodels.stats.contingency_tables.Table2x2.html#statsmodels.stats.contingency_tables.Table2x2"><code class="docutils literal notranslate"><span class="pre">statsmodels.Table2x2</span></code></a>:</p>
<ul>
<li><p>Specialized for 2x2 tables: as the name suggests, this class is specifically tailored for analyzing 2x2 contingency tables (tables with two rows and two columns).</p></li>
<li><p>Focused functionality: it provides a set of methods optimized for 2x2 tables, including:</p>
<ul>
<li><p>Odds ratio and relative risk estimation</p></li>
<li><p>Confidence intervals for OR, RR, and risk difference</p></li>
<li><p>McNemar’s test for paired data</p></li>
<li><p>Various tests of marginal homogeneity</p></li>
</ul>
</li>
<li><p>Convenience: it offers a streamlined interface for common analyses on 2x2 tables.</p></li>
<li><p>When to use: choose <code class="docutils literal notranslate"><span class="pre">Table2x2</span></code> when we have a 2x2 contingency table and need to perform the specific analyses it offers.</p></li>
</ul>
</li>
</ul>
</section>
<section id="interpreting-the-results-of-the-cochran-armitage-test">
<h4>Interpreting the results of the Cochran-Armitage test<a class="headerlink" href="#interpreting-the-results-of-the-cochran-armitage-test" title="Link to this heading">#</a></h4>
<p>The choice of scores can influence the test results. In <code class="docutils literal notranslate"><span class="pre">statsmodels.Table.test_ordinal_association</span></code>, the <code class="docutils literal notranslate"><span class="pre">col_scores</span></code> argument allows to specify the scores assigned to each column (category) of the contingency table. The direction of the trend (increasing or decreasing) is determined by the order of the scores. If the scores increase from left to right, the test will assess if there’s an increasing trend (default method behavior). If the scores decrease from left to right, the test will look for a decreasing trend.</p>
<p>In the context of a toxicology study, where we’re interested in the potential harmful effects of a treatment, a significant <em>increasing trend</em> in tumor rates (one-sided test) with increasing dose is a concerning finding. It suggests that the treatment might be promoting tumor development, and further investigation is needed to assess the safety and potential risks associated with the treatment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.contingency_tables</span> <span class="kn">import</span> <span class="n">Table</span>

<span class="c1"># Sample data: Tumor counts (rows) by Dose groups (columns)</span>
<span class="n">dose_groups</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Control&#39;</span><span class="p">,</span> <span class="s1">&#39;Low&#39;</span><span class="p">,</span> <span class="s1">&#39;Medium&#39;</span><span class="p">,</span> <span class="s1">&#39;High&#39;</span><span class="p">]</span>
<span class="n">tumor_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">],</span>  <span class="c1"># Control group</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">49</span><span class="p">],</span>  <span class="c1"># Low dose (1 tumor out of 50 total animals)</span>
    <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">47</span><span class="p">],</span>  <span class="c1"># Medium dose</span>
    <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">44</span><span class="p">]</span>   <span class="c1"># High dose</span>
<span class="p">])</span>

<span class="c1"># Perform Cochran-Armitage Trend Test</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">Table</span><span class="p">(</span><span class="n">tumor_counts</span><span class="p">,</span> <span class="n">shift_zeros</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">test_ordinal_association</span><span class="p">(</span>
    <span class="c1"># row_scores=np.array([0, 1, 2, 3]),  # default</span>
    <span class="n">col_scores</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]),</span>  <span class="c1"># reversed trend</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="c1"># Extract P value</span>
<span class="n">p_value</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">pvalue</span> <span class="o">/</span> <span class="mi">2</span>  <span class="c1"># Divide p-value by 2 for one-sided test</span>

<span class="c1"># Interpret results</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="k">if</span> <span class="n">p_value</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Reject the null hypothesis (P = </span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">; one-sided test). </span><span class="se">\</span>
<span class="se">\n</span><span class="s2">There is a significant increasing trend in tumor rates with increasing dose.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Fail to reject the null hypothesis (P = </span><span class="si">{</span><span class="n">p_value</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">; one-sided test). </span><span class="se">\</span>
<span class="se">\n</span><span class="s2">There is not enough evidence to suggest a trend.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>null_mean   15.0
null_sd     3.454659660058306
pvalue      0.0037959220933673713
statistic   25.0
zscore      2.894641146743591

Reject the null hypothesis (P = 0.002; one-sided test).
There is a significant increasing trend in tumor rates with increasing dose.
</pre></div>
</div>
</div>
</div>
<p>Be aware that the <code class="docutils literal notranslate"><span class="pre">statsmodels.Table</span></code> class, used for analyzing contingency tables, has a default behavior of adding 0.5 to <em>zero cell counts</em> if the <code class="docutils literal notranslate"><span class="pre">shift_zeros</span></code> argument is set to <code class="docutils literal notranslate"><span class="pre">True</span></code> (which is the default). This is done to avoid potential issues with zero frequencies when calculating statistics like the chi-squared test statistic. However, other statistical software or functions might not have this default behavior, potentially leading to different results if they encounter zero frequencies.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="n">Table</span><span class="p">(</span>
        <span class="n">tumor_counts</span><span class="p">,</span>
    <span class="p">)</span><span class="o">.</span><span class="n">test_ordinal_association</span><span class="p">(</span>
        <span class="c1"># row_scores=np.array([0, 1, 2, 3]),  # default</span>
        <span class="n">col_scores</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>  <span class="c1"># reversed trend</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>null_mean   15.71072319201995
null_sd     3.5390390971784127
pvalue      0.008669937476153359
statistic   25.0
zscore      2.624801973899118
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">test_ordinal_association</span></code> method from <code class="docutils literal notranslate"><span class="pre">statsmodels.stats.contingency_tables.Table</span></code> in Python performs the Cochran-Armitage test for trend. This test assesses whether there is a linear trend in the proportions of observations across ordered categories. It relies on an <em>asymptotic approximation</em> using the chi-squared distribution, which is generally accurate for large sample sizes.</p>
<p>For more sophisticated trend analysis or situations with small sample sizes, <a class="reference external" href="https://rcompanion.org/handbook/H_09.html">alternative methods are available in R</a>. For example, the <code class="docutils literal notranslate"><span class="pre">coin</span></code> package provides a flexible framework for permutation-based tests, including the linear-by-linear association test (<code class="docutils literal notranslate"><span class="pre">lbl_test</span></code>) and other trend tests. These permutation tests can be more accurate when the assumptions of the chi-squared test are not met or when dealing with small sample sizes. We can also use different scoring systems to test for linear, quadratic, or other types of trends.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">coin</span><span class="p">)</span>  <span class="c1"># https://cran.r-project.org/web/packages/coin/index.html</span>

<span class="n">table</span> <span class="o">&lt;-</span> <span class="n">matrix</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">50</span><span class="p">,</span><span class="mi">49</span><span class="p">,</span><span class="mi">47</span><span class="p">,</span><span class="mi">44</span><span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span> <span class="n">nrow</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">dimnames</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">resp</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="n">dose</span><span class="o">=</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">))</span>

<span class="n">lbl_test</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	Asymptotic Linear-by-Linear Association Test

data:  dose (ordered) by resp (1, 0)
Z = 2.8946, p-value = 0.003796
alternative hypothesis: two.sided
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform the Cochran-Armitage test using independence_test with &#39;exact&#39; distribution</span>
<span class="n">coin</span><span class="p">::</span><span class="n">independence_test</span><span class="p">(</span>
    <span class="k">as</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">table</span><span class="p">),</span>
    <span class="c1"># distribution = &#39;asymptotic&#39;,  # same as lbl_test()</span>
    <span class="n">distribution</span> <span class="o">=</span> <span class="s2">&quot;exact&quot;</span><span class="p">,</span>
    <span class="n">ytrafo</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">trafo</span><span class="p">(</span>
            <span class="n">data</span><span class="p">,</span>
            <span class="n">factor_trafo</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">model</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="o">~</span> <span class="n">x</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%*%</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	Exact General Independence Test

data:  dose by resp (1, 0)
Z = 2.8946, p-value = 0.004433
alternative hypothesis: two.sided
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="case-control-studies">
<h2>Case-control studies<a class="headerlink" href="#case-control-studies" title="Link to this heading">#</a></h2>
<p>In the previous sections, we’ve explored methods for analyzing data from cross-sectional and prospective studies, where we often have access to incidence rates or risk information directly. Case-control studies, however, present a different paradigm. These studies are <strong>retrospective</strong>, meaning they start with individuals who already have the outcome of interest (cases) and those who don’t (controls). The goal is to look back in time to assess whether exposure to a particular risk factor differs between the two groups. Though, they are prone to certain biases, such as recall bias (cases might remember their exposure history differently than controls) and selection bias (the way cases and controls are selected might not be representative of the underlying population).</p>
<section id="statistical-analysis-in-case-control-studies">
<h3>Statistical analysis in case-control studies<a class="headerlink" href="#statistical-analysis-in-case-control-studies" title="Link to this heading">#</a></h3>
<p>Unlike cohort studies, case-control studies are designed retrospectively, starting with individuals who already have the disease (cases) and those who don’t (controls). This structure means that the proportion of cases in the study doesn’t reflect the true disease prevalence in the population. Consequently, we cannot directly calculate risks or relative risks (RR) from case-control data.</p>
<p>Instead, case-control studies utilize the odds ratio (OR) as their primary measure of association. The OR compares the odds of exposure (e.g., vaccination) among cases to the odds of exposure among controls.  Mathematically, it’s the ratio of the odds of cases being exposed to the odds of controls being exposed.</p>
<p>The statistical tools we’ve discussed so far are still relevant, but their interpretation and application need to be adapted for case-control designs:</p>
<ul class="simple">
<li><p>Fisher’s exact test can still be used in case-control studies, especially when the sample size is <em>small</em> or the contingency table has low expected frequencies. However, it’s important to remember that Fisher’s test provides a P value for testing the null hypothesis of no association, but it doesn’t directly estimate the relative risk.</p></li>
<li><p>For larger sample sizes, the chi-squared test can be used as an alternative to Fisher’s exact test for testing the association between the exposure and outcome in a case-control study. Again, the chi-squared test doesn’t directly provide the RR, but it does give a P value to assess statistical significance.</p></li>
<li><p>The Cochran-Armitage test is generally not applicable to case-control studies because it’s designed for prospective studies where we follow individuals over time to observe the development of an outcome. In case-control studies, we’re looking back in time at exposure history, not tracking the development of new cases.</p></li>
</ul>
</section>
<section id="cholera-vaccination-example">
<h3>Cholera vaccination example<a class="headerlink" href="#cholera-vaccination-example" title="Link to this heading">#</a></h3>
<p>Case-control studies offer a powerful tool in epidemiology to explore the relationship between potential risk factors (exposures) and disease outcomes. They are particularly useful when studying rare diseases or outcomes with long latency periods, as they are often more efficient than prospective cohort studies.</p>
<p>The results of a case-control study are typically organized into a contingency table, where rows represent the outcome (case or control), columns represent the exposure (vaccinated or not vaccinated), and cell values represent the number of individuals in each category combination.</p>
<p>Let’s consider a classic example: investigating whether cholera vaccination is associated with a reduced risk of developing cholera, like in the <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/15728808/">study on oral cholera vaccination from Lucas and colleagues</a>. In this case-control study:</p>
<ul class="simple">
<li><p>Cases: individuals who have contracted cholera.</p></li>
<li><p>Controls: individuals from the same population who have not contracted cholera.</p></li>
</ul>
<p>We would then determine whether each individual in both groups had been previously vaccinated against cholera. The data could be organized in a contingency table like the one we saw earlier:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Vaccine</p></th>
<th class="head"><p>Cases (cholera)</p></th>
<th class="head"><p>Controls</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>10</p></td>
<td><p>94</p></td>
</tr>
<tr class="row-odd"><td><p>No</p></td>
<td><p>33</p></td>
<td><p>78</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create the contingency table from the data</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">94</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">33</span><span class="p">,</span> <span class="mi">78</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">Table2x2</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="c1"># Calculate the odds ratio (OR) and the 95% confidence interval</span>
<span class="n">oddsratio</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">oddsratio</span>
<span class="n">ci</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">oddsratio_confint</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Odds Ratio (OR): </span><span class="si">{</span><span class="n">oddsratio</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;95% Confidence Interval for OR: </span><span class="si">{</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bound</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">bound</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">ci</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Odds Ratio (OR): 0.251
95% Confidence Interval for OR: (0.117, 0.542)
</pre></div>
</div>
</div>
</div>
<p>The odds ratio quantifies the association between the vaccine (exposure) and the case/control status (outcome).</p>
<ul class="simple">
<li><p>An OR greater than 1 indicates that the exposure (vaccine) is associated with higher odds of being a case (having the disease).</p></li>
<li><p>An OR less than 1 indicates that the exposure (vaccine) is associated with lower odds of being a case (having the disease), suggesting a protective effect.</p></li>
<li><p>An OR of 1 means there’s no association between the exposure and the outcome.</p></li>
</ul>
<p>In our example, the OR of 0.25 signifies that vaccinated individuals have only 25% of the odds of developing the disease compared to unvaccinated individuals. This translates to a 75% reduction in the odds of disease due to vaccination.</p>
<p>Interestingly, the odds ratio often serves as a good approximation of the RR in many scenarios. This approximation is particularly accurate when the disease is rare. However, recent research suggests that the OR can be reasonably interpreted as the RR in most case-control studies, even for common diseases, with some caveats and careful consideration of the study design.</p>
<p>Moreover, the 95% confidence interval provides a range within which we can be 95% confident that the true population odds ratio lies. If the CI does not include 1, it indicates a <em>statistically significant</em> association at the 5% significance level.</p>
<p>Given the relatively small sample sizes in this case-control study, we might consider performing Fisher’s exact test in addition to calculating the OR and CI. This test is more appropriate for small samples and will provide an exact P value for the association.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">oddsratio</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Fisher&#39;s Exact Test P value:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p_value</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fisher&#39;s Exact Test P value: 0.0003
</pre></div>
</div>
</div>
</div>
</section>
<section id="isotretinoin-and-bowel-disease">
<h3>Isotretinoin and bowel disease<a class="headerlink" href="#isotretinoin-and-bowel-disease" title="Link to this heading">#</a></h3>
<p>Let’s use the example data on isotretinoin use and inflammatory bowel disease (IBD) in the <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/19623167/">case-control study from Bernstein and colleagues</a> to demonstrate a scenario where the confidence interval for the OR includes 1, indicating a lack of statistically significant association.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Isotretinoin</p></th>
<th class="head"><p>Cases (IBD)</p></th>
<th class="head"><p>Controls</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>25</p></td>
<td><p>213</p></td>
</tr>
<tr class="row-odd"><td><p>No</p></td>
<td><p>1935</p></td>
<td><p>19216</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create the contingency table from the data</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">213</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1935</span><span class="p">,</span> <span class="mi">19216</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">Table2x2</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">table</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>               Estimate   SE   LCB    UCB  p-value
--------------------------------------------------
Odds ratio        1.166        0.768 1.769   0.471
Log odds ratio    0.153 0.213 -0.264 0.570   0.471
Risk ratio        1.148        0.791 1.668   0.468
Log risk ratio    0.138 0.190 -0.235 0.511   0.468
--------------------------------------------------
</pre></div>
</div>
</div>
</div>
<p>The OR of 1.149 suggests that individuals exposed to isotretinoin have 1.149 times the odds of having IBD compared to those not exposed. However, this value is close to 1, indicating a weak association. Moreover, the 95% CI for the OR ranges from 0.768 to 1.769. Importantly, this interval includes 1. This means that we cannot rule out the possibility of no association between isotretinoin use and IBD in the population. We fail to reject the null hypothesis of no association at the 5% significance level.</p>
<p>Given the large sample size in this example, the chi-squared test could also be used to assess the association. However, Fisher’s exact test might be preferred if any of the cell counts were small.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">chi2</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">dof</span><span class="p">,</span> <span class="n">ex</span> <span class="o">=</span> <span class="n">chi2_contingency</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">correction</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chi-squared test P value:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Chi-squared test P value: 0.471
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">oddsratio</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Fisher&#39;s exact test P value:&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">p_value</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span> <span class="c1"># type: ignore</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fisher&#39;s exact test P value: 0.431
</pre></div>
</div>
</div>
</div>
<p>While the association is not statistically significant, it’s important to consider the potential clinical implications of even a small increased risk, especially for a drug with known side effects. Further research with larger sample sizes or different study designs might be warranted to clarify the relationship. The lack of statistical significance could be due to various factors, including limited sample size, potential biases in the study design, or the presence of confounding variables.</p>
</section>
</section>
<section id="conclusion">
<h2>Conclusion<a class="headerlink" href="#conclusion" title="Link to this heading">#</a></h2>
<p>In this chapter, we’ve embarked on a journey through the world of comparing proportions—a cornerstone of biostatistical analysis. We’ve explored a wide array of tools and techniques, from foundational contingency tables to powerful statistical tests, each designed to unveil patterns and associations hidden within categorical data.</p>
<p>We began with Fisher’s exact test, a precise method for analyzing small sample sizes and rare events. We then delved into the versatile chi-squared test, capable of handling larger tables and assessing deviations from expected frequencies. We further refined our toolkit with the Cochran-Armitage trend test, designed to detect trends in ordinal data, particularly useful in toxicological studies.</p>
<p>Along the way, we explored essential concepts like attributable risk, relative risk, odds ratio, and number needed to treat, equipping us with the language and understanding to quantify the impact of exposures and treatments. We learned how to construct confidence intervals for these measures, acknowledging and embracing the uncertainty inherent in statistical estimation.</p>
<p>Finally, we turned our attention to case-control studies, a powerful observational design for exploring associations in retrospective data. We discovered the importance of the odds ratio in this context and how it can approximate the relative risk under certain conditions.</p>
<p>Throughout this chapter, we’ve emphasized the importance of selecting the right statistical tool for the job. The choice of test depends on the study design, data characteristics, and research questions. While some tests are more versatile than others, each has its own strengths and limitations. Understanding these nuances is crucial for conducting rigorous and reliable analyses. For example, smaller sample sizes might require exact tests (like Fisher’s) or adjustments (like Yates’ correction), while larger samples can often rely on approximations (like the chi-squared test).</p>
<p>We’ve learned how to harness the power of Python to perform these analyses, using libraries like <code class="docutils literal notranslate"><span class="pre">scipy.stats</span></code> and <code class="docutils literal notranslate"><span class="pre">statsmodels</span></code>. We’ve seen how these tools can be applied to real-world problems, from assessing the efficacy of treatments to analyzing dose-response relationships in toxicology studies.</p>
<p>As we conclude this chapter, it’s crucial to remember that statistics are more than just numbers. They are lenses through which we interpret data, helping us make sense of complex phenomena and inform decision-making. However, statistical results should always be interpreted in context.</p>
</section>
<section id="cheat-sheet">
<h2>Cheat sheet<a class="headerlink" href="#cheat-sheet" title="Link to this heading">#</a></h2>
<p>This cheat sheet provides a quick reference for essential code snippets used in this chapter.</p>
<section id="id1">
<h3>Contingency table<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">contingency</span>

<span class="c1"># Example data</span>
<span class="n">treatment</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">outcome</span>   <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>

<span class="c1"># Create contingency table with scipy.stats.contingency</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">contingency</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">treatment</span><span class="p">,</span> <span class="n">outcome</span><span class="p">)</span> <span class="c1"># treatment will be the rows, outcome the columns</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">count</span>

<span class="c1"># If the data is already in a Pandas DataFrame we can use pandas.crosstab</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1"># Assuming the Dataframe has columns named &quot;treatment&quot; and &quot;outcome&quot;</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;treatment&quot;</span><span class="p">:</span> <span class="n">treatment</span><span class="p">,</span> <span class="s2">&quot;outcome&quot;</span><span class="p">:</span> <span class="n">outcome</span><span class="p">})</span>

<span class="c1"># Creating contingency table using pd.crosstab</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;treatment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">(),</span>  <span class="c1"># Remove NaN values if expected</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;outcome&quot;</span><span class="p">],</span>
    <span class="n">margins</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>     <span class="c1"># Include row and column totals</span>
    <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span>      <span class="c1"># Keep rows/columns with zero counts</span>
<span class="p">)</span>

<span class="c1"># # Pre-aggregated contingency table with rows = alternative treatments,</span>
<span class="c1"># cols = alternative outcomes; here is the example from Apixaban study</span>
<span class="n">table_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s1">&#39;recurrence&#39;</span><span class="p">:</span>    <span class="p">[</span><span class="mi">73</span><span class="p">,</span> <span class="mi">14</span><span class="p">],</span>
        <span class="s1">&#39;no_recurrence&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">756</span><span class="p">,</span> <span class="mi">826</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;placebo&#39;</span><span class="p">,</span> <span class="s1">&#39;apixaban&#39;</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="id2">
<h3>Fisher’s exact test<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">fisher_exact</span>

<span class="c1"># Assuming &#39;table&#39; is (pre-aggregated) 2x2 contingency table (NumPy array or list of lists)</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span> <span class="c1"># converts DataFrame to NumPy array</span>
<span class="n">odds_ratio</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">fisher_exact</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;two-sided&#39;</span><span class="p">)</span>
<span class="c1"># could use &quot;alternative=&#39;greater&#39;&quot; for toxicology count data</span>
</pre></div>
</div>
</section>
<section id="risk-and-odds-ratio-analysis">
<h3>Risk and odds ratio analysis<a class="headerlink" href="#risk-and-odds-ratio-analysis" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.contingency_tables</span> <span class="kn">import</span> <span class="n">Table2x2</span>

<span class="c1"># Assuming &#39;table&#39; is the 2x2 contingency table (NumPy array)</span>
<span class="n">table2x2</span> <span class="o">=</span> <span class="n">Table2x2</span><span class="p">(</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="c1"># Calculate and print the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">table2x2</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>

<span class="c1"># Calculates probabilities</span>
<span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">p1</span><span class="p">,</span> <span class="n">p2</span> <span class="o">=</span> <span class="n">a</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">),</span> <span class="n">c</span> <span class="o">/</span> <span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>
<span class="c1"># Confidence intervals</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.proportion</span> <span class="kn">import</span> <span class="n">proportion_confint</span>
<span class="n">conf</span><span class="o">=</span><span class="mf">0.95</span>  <span class="c1"># Set 95% confidence interval</span>
<span class="n">ci_p1</span> <span class="o">=</span> <span class="n">proportion_confint</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">nobs</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">conf</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">)</span>
<span class="n">ci_p2</span> <span class="o">=</span> <span class="n">proportion_confint</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="n">c</span><span class="p">,</span> <span class="n">nobs</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">-</span><span class="n">conf</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">)</span>

<span class="c1"># Overall risk</span>
<span class="n">p_bar</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">c</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span> <span class="o">+</span> <span class="n">c</span> <span class="o">+</span> <span class="n">d</span><span class="p">)</span>

<span class="c1"># Calculates absolute risk reduction / attributable risk</span>
<span class="n">arr</span> <span class="o">=</span> <span class="n">p1</span> <span class="o">-</span> <span class="n">p2</span>
<span class="c1"># Confidence interval estimation</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.proportion</span> <span class="kn">import</span> <span class="n">confint_proportions_2indep</span>
<span class="n">ci_arr_1</span><span class="p">,</span> <span class="n">ci_arr_2</span> <span class="o">=</span> <span class="n">confint_proportions_2indep</span><span class="p">(</span>
    <span class="n">count1</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">nobs1</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">count2</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">nobs2</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span>
    <span class="n">compare</span><span class="o">=</span><span class="s1">&#39;diff&#39;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;score&#39;</span><span class="p">,</span>  <span class="c1"># [&#39;wald&#39;, &#39;agresti-caffo&#39;, &#39;newcomb&#39;, &#39;score&#39;]</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>

<span class="c1"># Calculates number needed to treat</span>
<span class="n">ntt</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="n">arr</span>
<span class="c1"># Confidence interval is the reciprocal of ARR CI</span>
<span class="n">ci_ntt_1</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">ci_arr_2</span><span class="p">)</span>
<span class="n">ci_ntt_2</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">ci_arr_1</span><span class="p">)</span>

<span class="c1"># Calculates relative risk</span>
<span class="n">rr</span> <span class="o">=</span> <span class="n">p1</span><span class="o">/</span><span class="n">p2</span>
<span class="c1"># Alternatively using scipy.stats.contingency</span>
<span class="n">contingency</span><span class="o">.</span><span class="n">relative_risk</span><span class="p">(</span>
    <span class="n">exposed_cases</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">exposed_total</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">control_cases</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">control_total</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span>
<span class="p">)</span><span class="o">.</span><span class="n">relative_risk</span>
<span class="c1"># Confidence interval estimation</span>
<span class="n">ci_rr_1</span><span class="p">,</span> <span class="n">ci_rr_2</span> <span class="o">=</span> <span class="n">confint_proportions_2indep</span><span class="p">(</span>
    <span class="n">count1</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">nobs1</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">count2</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">nobs2</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span>
    <span class="n">compare</span><span class="o">=</span><span class="s1">&#39;ratio&#39;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;score&#39;</span><span class="p">,</span>  <span class="c1"># [&#39;log&#39;, &#39;log-adjusted&#39;, &#39;score&#39;]</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>

<span class="c1"># Calculates odds ratio</span>
<span class="ow">or</span> <span class="o">=</span> <span class="n">a</span><span class="o">*</span><span class="n">d</span> <span class="o">/</span> <span class="p">(</span><span class="n">b</span><span class="o">*</span><span class="n">c</span><span class="p">)</span>
<span class="c1"># Alternatively using scipy.stats.contingency</span>
<span class="n">contingency</span><span class="o">.</span><span class="n">odds_ratio</span><span class="p">(</span>
    <span class="n">table</span><span class="o">=</span><span class="n">table_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span>
    <span class="n">kind</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">statistic</span>
<span class="c1"># Confidence interval estimation</span>
<span class="n">ci_or_1</span><span class="p">,</span> <span class="n">ci_or_2</span> <span class="o">=</span> <span class="n">confint_proportions_2indep</span><span class="p">(</span>
    <span class="n">count1</span><span class="o">=</span><span class="n">a</span><span class="p">,</span>
    <span class="n">nobs1</span><span class="o">=</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span>
    <span class="n">count2</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
    <span class="n">nobs2</span><span class="o">=</span><span class="n">c</span><span class="o">+</span><span class="n">d</span><span class="p">,</span>
    <span class="n">compare</span><span class="o">=</span><span class="s1">&#39;odds-ratio&#39;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;score&#39;</span><span class="p">,</span>  <span class="c1"># [&#39;logit&#39;, &#39;logit-adjusted&#39;, &#39;score&#39;]</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="z-test">
<h3>Z-test<a class="headerlink" href="#z-test" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.proportion</span> <span class="kn">import</span> <span class="n">proportions_ztest</span>

<span class="c1"># Extract counts and total observations for each group from the Apixaban table</span>
<span class="n">count</span> <span class="o">=</span> <span class="n">table_df</span><span class="p">[</span><span class="s1">&#39;recurrence&#39;</span><span class="p">]</span>
<span class="n">nobs</span> <span class="o">=</span> <span class="n">table_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Perform Z-test for difference in proportions</span>
<span class="n">z_stat</span><span class="p">,</span> <span class="n">p_value</span> <span class="o">=</span> <span class="n">proportions_ztest</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">nobs</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id3">
<h3>Chi-squared goodness-of-fit test<a class="headerlink" href="#id3" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chisquare</span>

<span class="c1"># Mendel&#39;s observed data</span>
<span class="n">observed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">315</span><span class="p">,</span> <span class="mi">108</span><span class="p">,</span> <span class="mi">101</span><span class="p">,</span> <span class="mi">32</span><span class="p">])</span>

<span class="c1"># Mendel&#39;s expected proportions (here as fractions)</span>
<span class="n">expected_proportions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">9</span><span class="o">/</span><span class="mi">16</span><span class="p">,</span> <span class="mi">3</span><span class="o">/</span><span class="mi">16</span><span class="p">,</span> <span class="mi">3</span><span class="o">/</span><span class="mi">16</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="mi">16</span><span class="p">])</span>

<span class="c1"># Perform the chi-squared goodness-of-fit test</span>
<span class="n">chisquare</span><span class="p">(</span>
    <span class="n">f_obs</span><span class="o">=</span><span class="n">observed</span><span class="p">,</span>
    <span class="n">f_exp</span><span class="o">=</span><span class="n">observed</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">*</span> <span class="n">expected_proportions</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="chi-squared-distribution">
<h3>Chi-squared distribution<a class="headerlink" href="#chi-squared-distribution" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chi2</span>

<span class="c1"># Define the standard normal distribution model</span>
<span class="n">df</span> <span class="o">=</span> <span class="mi">5</span>
<span class="n">chi2_distribution</span> <span class="o">=</span> <span class="n">chi2</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="p">)</span>
<span class="c1"># Distribution statistics</span>
<span class="n">chi2_distribution</span><span class="o">.</span><span class="n">stats</span><span class="p">()</span>

<span class="c1"># Probabilities</span>
<span class="c1"># Cumulative Distribution Function - P(X&lt;=8)</span>
<span class="n">chi2_distribution</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="c1"># (One-tailed) Survival Function - P(X&gt;=16)</span>
<span class="n">chi2_distribution</span><span class="o">.</span><span class="n">sf</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
<span class="c1"># Percent Point Function</span>
<span class="n">chi2_distribution</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="n">q</span><span class="o">=</span><span class="mf">.95</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id4">
<h3>Chi-squared test of independence<a class="headerlink" href="#id4" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Mendel&#39;s observed data (2x2 contingency table)</span>
<span class="n">observed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="p">[</span><span class="mi">315</span><span class="p">,</span> <span class="mi">101</span><span class="p">],</span>  <span class="c1"># Yellow, round and wrinkled</span>
        <span class="p">[</span><span class="mi">108</span><span class="p">,</span> <span class="mi">32</span><span class="p">]</span>    <span class="c1"># Green,  round and wrinkled</span>
    <span class="p">])</span>

<span class="c1"># Create a DataFrame from the observed data</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">observed</span><span class="p">,</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Round&#39;</span><span class="p">,</span> <span class="s1">&#39;Wrinkled&#39;</span><span class="p">],</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Yellow&#39;</span><span class="p">,</span> <span class="s1">&#39;Green&#39;</span><span class="p">])</span>

<span class="c1"># Create a mosaic plot</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">statsmodels.graphics.mosaicplot</span> <span class="kn">import</span> <span class="n">mosaic</span>
<span class="n">mosaic</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">stack</span><span class="p">())</span>  <span class="c1"># Stacking the DataFrame for proper mosaic plot input</span>

<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chi2_contingency</span>

<span class="c1"># Perform the chi-squared test of independence on the previous contingency table</span>
<span class="n">chi2_contingency</span><span class="p">(</span>
    <span class="n">observed</span><span class="o">=</span><span class="n">observed</span><span class="p">,</span>
    <span class="n">lambda_</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">,</span>  <span class="c1"># [&#39;log-likelihood&#39;, &#39;freeman-tukey&#39;, &#39;neyman&#39;, &#39;cressie-read&#39;]</span>
    <span class="n">correction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>    <span class="c1"># Yates&#39; correction for continuity</span>
<span class="p">)</span>

<span class="c1"># Chi-squared independence tests using pingouin</span>
<span class="kn">import</span> <span class="nn">pingouin</span> <span class="k">as</span> <span class="nn">pg</span>
<span class="n">pg</span><span class="o">.</span><span class="n">chi2_independence</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">table_df</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;recurrence&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;no_recurrence&#39;</span><span class="p">,</span>
    <span class="n">correction</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id5">
<h3>Cochran-Armitage test for trend<a class="headerlink" href="#id5" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.stats.contingency_tables</span> <span class="kn">import</span> <span class="n">Table</span>

<span class="c1"># Sample data: Tumor counts (rows) by Dose groups (columns)</span>
<span class="n">dose_groups</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Control&#39;</span><span class="p">,</span> <span class="s1">&#39;Low&#39;</span><span class="p">,</span> <span class="s1">&#39;Medium&#39;</span><span class="p">,</span> <span class="s1">&#39;High&#39;</span><span class="p">]</span>
<span class="n">tumor_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">],</span>  <span class="c1"># Control group</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">49</span><span class="p">],</span>  <span class="c1"># Low dose (1 tumor out of 50 total animals)</span>
    <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">47</span><span class="p">],</span>  <span class="c1"># Medium dose</span>
    <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">44</span><span class="p">]</span>   <span class="c1"># High dose</span>
<span class="p">])</span>

<span class="c1"># Perform Cochran-Armitage trend test (asymptotic linear-by-linear)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="n">Table</span><span class="p">(</span>
        <span class="n">tumor_counts</span><span class="p">,</span>
        <span class="n">shift_zeros</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span><span class="o">.</span><span class="n">test_ordinal_association</span><span class="p">(</span>
        <span class="c1"># row_scores=np.array([0, 1, 2, 3]), # default</span>
        <span class="n">col_scores</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>   <span class="c1"># reversed trend</span>
    <span class="p">))</span>
<span class="c1"># Divide P value by 2 for one-sided test</span>
</pre></div>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">6</span><span class="p">,</span><span class="m">50</span><span class="p">,</span><span class="m">49</span><span class="p">,</span><span class="m">47</span><span class="p">,</span><span class="m">44</span><span class="p">),</span><span class="w"> </span><span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">resp</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">dose</span><span class="o">=</span><span class="m">0</span><span class="o">:</span><span class="m">3</span><span class="p">))</span>

<span class="c1"># https://cran.r-project.org/web/packages/coin/index.html</span>
<span class="nf">library</span><span class="p">(</span><span class="n">coin</span><span class="p">)</span>
<span class="nf">lbl_test</span><span class="p">(</span><span class="nf">as.table</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>

<span class="n">coin</span><span class="o">::</span><span class="nf">independence_test</span><span class="p">(</span>
<span class="w">    </span><span class="nf">as.table</span><span class="p">(</span><span class="n">table</span><span class="p">),</span>
<span class="w">    </span><span class="c1"># distribution = &#39;asymptotic&#39;,  # same as lbl_test()</span>
<span class="w">    </span><span class="n">distribution</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;exact&#39;</span><span class="p">,</span>
<span class="w">    </span><span class="n">ytrafo</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="nf">trafo</span><span class="p">(</span>
<span class="w">            </span><span class="n">data</span><span class="p">,</span>
<span class="w">            </span><span class="n">factor_trafo</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">colnames</span><span class="p">(</span><span class="n">table</span><span class="p">))}</span>
<span class="w">        </span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id6">
<h3>Case-control studies<a class="headerlink" href="#id6" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create the contingency table from the data</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">213</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1935</span><span class="p">,</span> <span class="mi">19216</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">Table2x2</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">table</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>

<span class="n">chi2_contingency</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="session-information">
<h2>Session information<a class="headerlink" href="#session-information" title="Link to this heading">#</a></h2>
<p>The output below details all packages and version necessary to reproduce the results in this report.</p>
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>python<span class="w"> </span>--version
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-------------&quot;</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">importlib.metadata</span> <span class="kn">import</span> <span class="n">version</span>

<span class="c1"># List of packages we want to check the version</span>
<span class="n">packages</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;numpy&#39;</span><span class="p">,</span> <span class="s1">&#39;pandas&#39;</span><span class="p">,</span> <span class="s1">&#39;matplotlib&#39;</span><span class="p">,</span> <span class="s1">&#39;seaborn&#39;</span><span class="p">,</span> <span class="s1">&#39;scipy&#39;</span><span class="p">,</span> <span class="s1">&#39;pingouin&#39;</span><span class="p">,</span> <span class="s1">&#39;statsmodels&#39;</span><span class="p">]</span>

<span class="c1"># Initialize an empty list to store the versions</span>
<span class="n">versions</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Loop over the packages</span>
<span class="k">for</span> <span class="n">package</span> <span class="ow">in</span> <span class="n">packages</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Get the version of the package</span>
        <span class="n">package_version</span> <span class="o">=</span> <span class="n">version</span><span class="p">(</span><span class="n">package</span><span class="p">)</span>
        <span class="c1"># Append the version to the list</span>
        <span class="n">versions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">package_version</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>  <span class="c1"># Use a more general exception for broader compatibility</span>
        <span class="n">versions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Not installed&#39;</span><span class="p">)</span>

<span class="c1"># Print the versions</span>
<span class="k">for</span> <span class="n">package</span><span class="p">,</span> <span class="n">version</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">packages</span><span class="p">,</span> <span class="n">versions</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">package</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Python 3.12.7
-------------
numpy: 1.26.4
pandas: 2.2.2
matplotlib: 3.9.2
seaborn: 0.13.2
scipy: 1.14.1
pingouin: 0.5.5
statsmodels: 0.14.2
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sessionInfo</span><span class="p">()</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">NameError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">68</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">sessionInfo</span><span class="p">()</span>

<span class="ne">NameError</span>: name &#39;sessionInfo&#39; is not defined
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="24%20-%20Normality%20tests%20and%20outliers.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Normality tests and outliers</p>
      </div>
    </a>
    <a class="right-next"
       href="29%20-%20Comparing%20survival%20curves.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Comparing survival curves</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#fisher-s-exact-test">Fisher’s exact test</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#when-to-use-fisher-s-test">When to use Fisher’s test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#contingency-table">Contingency table</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#p-value">P value</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#hypergeometric-distribution">Hypergeometric distribution</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#relationship-between-hypergeometric-distribution-and-p-value">Relationship between hypergeometric distribution and P value</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-p-value">Interpreting the P value</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-and-interpreting-risk-metrics">Calculating and interpreting risk metrics</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#absolute-risk-reduction-arr-and-attributable-risk-ar">Absolute risk reduction (ARR) and attributable risk (AR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-attributable-risk-par">Population attributable risk (PAR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#number-needed-to-treat-ntt">Number needed to treat (NTT)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#relative-risk-rr">Relative risk (RR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#logrr">LogRR</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#odds-ratio-or">Odds ratio (OR)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#logor">LogOR</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#confidence-intervals">Confidence intervals</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-proportions">CI for proportions</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-arr">CI for ARR</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#standard-approximations">Standard approximations</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#z-test-for-comparing-proportions">Z-test for comparing proportions</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#exact-methods">Exact methods</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-nnt">CI for NNT</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-rr">CI for RR</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ci-for-or">CI for OR</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-table">Summary table</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#calculating-measures-of-association-and-effect-with-custom-functions">Calculating measures of association and effect with custom functions</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#table2x2-for-risk-and-odds-ratio-analysis">Table2x2 for risk and odds ratio analysis</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#one-sided-test-application-to-toxicology-count-data">One-sided test - Application to toxicology count data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-test">Chi-squared test</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-it-works">How it works</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-of-the-mendel-s-peas">Example of the Mendel’s peas</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-goodness-of-fit-test">Chi-squared goodness-of-fit test</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-chi-squared-distribution">The chi-squared distribution</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#degrees-of-freedom-in-goodness-of-fit-tests">Degrees of freedom in goodness-of-fit tests</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-test-of-independence">Chi-squared test of independence</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mosaic-plot">Mosaic plot</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-results-of-the-chi-squared-test-of-independence">Interpreting the results of the chi-squared test of independence</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#degrees-of-freedom-in-independence-tests">Degrees of freedom in independence tests</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#yates-correction">Yates’ correction</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#directional-hypothesis">Directional hypothesis</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-independence-tests-using-pingouin">Chi-squared independence tests using Pingouin</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cochran-armitage-test-for-trend">Cochran-Armitage test for trend</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-the-trend-test-works">How the trend test works</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-of-varying-the-dose-of-a-drug">Example of varying the dose of a drug</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#table-vs-table2x2"><code class="docutils literal notranslate"><span class="pre">Table</span></code> vs. <code class="docutils literal notranslate"><span class="pre">Table2x2</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-results-of-the-cochran-armitage-test">Interpreting the results of the Cochran-Armitage test</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#case-control-studies">Case-control studies</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-analysis-in-case-control-studies">Statistical analysis in case-control studies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cholera-vaccination-example">Cholera vaccination example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#isotretinoin-and-bowel-disease">Isotretinoin and bowel disease</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">Conclusion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cheat-sheet">Cheat sheet</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Contingency table</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Fisher’s exact test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#risk-and-odds-ratio-analysis">Risk and odds ratio analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#z-test">Z-test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Chi-squared goodness-of-fit test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chi-squared-distribution">Chi-squared distribution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Chi-squared test of independence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Cochran-Armitage test for trend</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Case-control studies</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-information">Session information</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sébastien Wieckowski
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>